---
title: "Figure 3"
author: "Ludvig Larsson"
date: '2022-05-23'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load libraries

```{r warning=FALSE, message=FALSE}

library(STutility)
library(ggplot2)
library(ggpubr)
library(magrittr)
library(dplyr)

```

# Download files from Mendeley Data 
***

Note that you only have to run this code chunk once. If you are having issue downloading files, please make sure that 
you have a stable internet connection and that Mendeley Data (https://data.mendeley.com/) is up and running.

These files should be downloaded without any issue. If you encounter any errors, some of the code in this 
notebook will be possible to run.

If you struggle to download the data from R, you can also download the files manually from Mendeley Data and
put them in a folder called `data/` at the root of the project.

```{r eval=FALSE}

# Create data directories
suppressWarnings({
  dir.create("../data/")
  dir.create("../data/genes/")
  dir.create("../data/sheets/")
  dir.create("../data/R_objects/")
  dir.create("../data/spaceranger_output/")
})

source("../scripts/mendeley.R")

# Download gene annotation files
for (i in seq_along(genes.files)) {
  download.file(url = genes.files[i], destfile = paste0("../data/genes/", names(genes.files)[i]))
}

# Download gene annotation files
for (i in seq_along(sheets.files)) {
  download.file(url = sheets.files[i], destfile = paste0("../data/sheets/", names(sheets.files)[i]))
}

# Download R object
download.file(url = r.files[1], destfile = paste0("../data/R_objects/", names(r.files)[1]))

# Download spaceranger output files
download.file(url = spaceranger.files[2], destfile = paste0("../data/spaceranger_output/", names(spaceranger.files)[2]))
untar(tarfile = paste0("../data/spaceranger_output/", names(spaceranger.files)[2]), exdir = "../data/spaceranger_output/")
file.remove(paste0("../data/spaceranger_output/", names(spaceranger.files)[2]))

```

Assemble spaceranger output files and merge curated meta data

```{r}

samples <- Sys.glob(paths = "../data/spaceranger_output/colon/*/filtered_feature_bc_matrix.h5")
imgs <- Sys.glob(paths = "../data/spaceranger_output/colon/*/spatial/tissue_hires_image.png")
spotfiles <- Sys.glob(paths = "../data/spaceranger_output/colon/*/spatial/tissue_positions_list.csv")
json <- Sys.glob(paths = "../data/spaceranger_output/colon/*/spatial/scalefactors_json.json")

infoTable <- data.frame(samples, imgs, spotfiles, json)
infoTable <- cbind(infoTable, arrayID = do.call(rbind, strsplit(infoTable$samples, "/"))[, 5])

curated_metadata <- openxlsx::read.xlsx("../data/sheets/RRST_sample_metadata.xlsx")
curated_metadata <- setNames(curated_metadata, nm = c("storage_time", "seq_date", "comments",
                                                       "ID", "paper_id", "RIN", "DV200", "protocol", "source", 
                                                       "arrayID", "spots_under_tissue",
                                                       "genes_detected", "fraction_spots_under_tissue",
                                                       "median_genes_per_spot", "median_UMIs_per_spot", 
                                                       "saturation", "reads_mapped_to_probe_set",
                                                       "reads_mapped_confidently_to_probe_set",
                                                       "reads_mapped_confidently_to_filtered_probe_set",
                                                       "reads_mapped_to_genome",
                                                       "reads_mapped_confidently_to_genome",
                                                       "number_of_panel_genes"))

infoTable <- merge(infoTable, curated_metadata, by = "arrayID")

```

Load data into a Seurat object

```{r}

CLN <- InputFromTable(infoTable[c(5:6, 2, 4, 1, 3), ])

```

Check spatial distribution of unique genes

```{r fig.width=10, fig.height=6}

CLN$protocol_array <- paste0(CLN$protocol, " : ", CLN$arrayID)
ST.FeaturePlot(CLN, features = "nFeature_RNA", ncol = 3, label.by = "protocol_array", show.sb = FALSE, pt.size = 1.5)

```

Load images

```{r}

CLN <- LoadImages(CLN, time.resolve = FALSE, xdim = 1e3)

```


Apply rigid transformations to make a rough alignment of the tissue sections.

```{r}

# Warp transform
CLN <- WarpImages(CLN, verbose = TRUE, transforms = list("1" = list(angle = -90), 
                                                         "2" = list(angle = -90, mirror.x = TRUE), 
                                                         "4" = list(angle = -90), 
                                                         "5" = list(angle = 180), 
                                                         "6" = list(angle = 180)))

```

## Add manual annotations
***

Here, the spots were labeled into three major categories: "mucosa", "submucosa" and "muscularis".

The annotations are provided in an Rds file (see below) so you can skip the next code chunk.

```{r eval=FALSE}

# Only run if you want to annotate the images
CLN <- ManualAnnotation(CLN)

```

```{r}

# Add labels
pre_ann <- readRDS("../data/R_objects/CLN_metadata_selections")
CLN@meta.data[rownames(pre_ann), "labels"] <- pre_ann[, "labels"]

```

Add a new metadata column with a combined protocol and lung ID label

```{r fig.width=12, fig.height=8}

CLN$protocol_sample <- gsub(pattern = " ", replacement = "_", paste0(CLN$protocol, "_ID: ", CLN$paper_id))
ST.FeaturePlot(CLN, features = "nFeature_RNA", ncol = 3, label.by = "protocol_sample")

```

Filter out background

```{r}

CLN <- SubsetSTData(CLN, expression = labels %in% c("mucosa", "submucosa", "muscularis"))

```

## Create H&E + spatial maps of unique genes for Figure 3a
***

```{r fig.width=12, fig.height=6}


infoTable[c(5:6, 2, 4, 1, 3), 1]

limits_list <- list("2" = c(x_start = 260, y_start = 500, x_end = 1600, y_end = 1800),
                    "3" = c(x_start = 170, y_start = 340, x_end = 1400, y_end = 1500),
                    "4" = c(x_start = 130, y_start = 240, x_end = 1780, y_end = 1560),
                    "6" = c(x_start = 300, y_start = 280, x_end = 1820, y_end = 1700))

ann_plots <- lapply(c("2", "3", "4", "6"), function(i) {
  
  gg <- cbind(CLN[[]], GetStaffli(CLN)@meta.data)
  gg <- subset(gg, sample == i)
  dims <- GetStaffli(CLN)@dims[[i]]
  
  p <- ggplot(gg, aes(warped_x, dims$height - warped_y, color = labels)) +
    geom_point(size = 1) +
    theme_void() +
    scale_color_manual(values = c("mucosa" = "#AA4499", "submucosa" = "#DDCC77", "muscularis" = "#CC6677")) +
    scale_x_continuous(expand = c(0, 0), limits = c(limits_list[[i]]["x_start"], 
                                  limits_list[[i]]["x_end"])) +
    scale_y_continuous(expand = c(0, 0), limits = c(dims$height - limits_list[[i]]["y_end"], 
                                  dims$height - limits_list[[i]]["y_start"])) +
    theme(legend.position = "none")

  p <- ggrastr::rasterize(p, layers = "Point", dpi = 300)
  return(p)
})

# HE images
plots_HE <- lapply(c("2", "3", "4", "6"), function(i) {
  im <- GetStaffli(CLN)@rasterlists$processed[[paste0(i)]]
  dims <- GetStaffli(CLN)@dims[[i]]
   
  im <- im[limits_list[[i]]["y_start"]:limits_list[[i]]["y_end"]*0.5, limits_list[[i]]["x_start"]:limits_list[[i]]["x_end"]*0.5]
  g <- grid::rasterGrob(im, width = unit(1, "npc"), height = unit(1, "npc"), interpolate = TRUE)
  p <- ggplot() +
    annotation_custom(g, -Inf, Inf, -Inf, Inf) +
    theme_void() +
    theme(plot.margin = margin(t = 0, r = 0, b = 0, l = 0))
  return(p)
})

p1 <- plots_HE[[1]] + 
  ann_plots[[1]]
p1

p2 <-  plots_HE[[2]] +
  ann_plots[[2]]
p2

p3 <-  plots_HE[[3]] +
  ann_plots[[3]]
p3

p4 <-  plots_HE[[4]] +
  ann_plots[[4]]
p4

```

```{r eval=FALSE, include=FALSE}

df_CLN_qc <- CLN[[]] |> 
  select(protocol, nFeature_RNA, nCount_RNA, paper_id, labels) |> 
  bind_cols(GetStaffli(CLN)@meta.data |> select(warped_x, warped_y)) |> 
  rename(unique_genes = nFeature_RNA, annotation = labels, UMIs = nCount_RNA)

openxlsx::write.xlsx(df_CLN_qc |> select(warped_x, warped_y, protocol, annotation, paper_id), file = "../../submission/source_data/source_data_3a.xlsx", sheetName = "3a")

openxlsx::write.xlsx(df_CLN_qc |> select(warped_x, warped_y, protocol, unique_genes, paper_id), file = "../../submission/source_data/source_data_3b.xlsx", sheetName = "3b")

openxlsx::write.xlsx(df_CLN_qc |> select(protocol, unique_genes, paper_id), file = "../../submission/source_data/source_data_3c.xlsx", sheetName = "3c")

openxlsx::write.xlsx(df_CLN_qc |> select(protocol, UMIs, paper_id), file = "../../submission/source_data/source_data_3d.xlsx", sheetName = "3d")

```


```{r eval=FALSE}

pdf(file = "plots/sample1_HE_unique_genes_RRST.pdf", width = 6.3, height = 3.05)
print(p1)
dev.off()

pdf(file = "plots/sample1_HE_unique_genes_standard.pdf", width = 5.7, height = 2.9)
print(p2)
dev.off()

pdf(file = "plots/sample2_HE_unique_genes_RRST.pdf", width = 7.8, height = 3.2)
print(p3)
dev.off()

pdf(file = "plots/sample2_HE_unique_genes_standard.pdf", width = 7, height = 3.4)
print(p4)
dev.off()

```



## Violin plots for Figure 3 c and d
***

```{r fig.width=14, fig.height=6, eval=FALSE}

th <- theme(panel.background = element_rect(fill = "white", colour = "lightgray"), 
          panel.grid = element_line(colour = "lightgray", linetype = "longdash"), 
          axis.text.x = element_blank(),
          axis.text.y = element_text(size = 14, color = "black"), 
          plot.title = element_text(size = 18, face = "bold"), 
          axis.ticks.x = element_blank(), 
          strip.text = element_text(size = 16, color = "white", face = "bold"),
          legend.text = element_text(size = 14),
          legend.title = element_text(size = 16), legend.position = "bottom") 

dt <- CLN[[]] %>% 
  mutate(labels = factor(setNames(c("Mucosa", "Submucosa", "Muscularis"), 
                                  nm = c("mucosa", "submucosa", "muscularis"))[labels], 
                         levels = c("Mucosa", "Submucosa", "Muscularis")), 
         arrayID = factor(arrayID, levels = c("V11M22-349_A1", "V11M22-349_B1", "V11A20-396_A1",
                                              "V10S29-108_B1", "V11B18-363_C1", "V11A20-396_C1")))

p1 <- ggplot() +
    geom_violin(data = dt, aes(arrayID, nFeature_RNA, fill = protocol), scale = "width") +
    geom_point(data = dt %>% 
                 mutate(replicate = setNames(c("rep1", "rep2", "rep1", "rep1", "rep2", "rep1"), 
                                       nm = c("V11M22-349_A1", "V11M22-349_B1", "V11A20-396_A1",
                                              "V10S29-108_B1", "V11B18-363_C1", "V11A20-396_C1"))[arrayID]) %>%
                 group_by(labels, paper_id, arrayID, replicate) %>% 
                 summarize(Mean = median(nFeature_RNA)), 
               aes(arrayID, Mean, shape = replicate), size = 3) +
    scale_shape_manual(values = c("rep1" = 4, "rep2" = 6)) +
    facet_grid(paper_id~labels, scales = "free", space = "free") +
    labs(y = "", x = "", title = "Unique genes") +
    th +
    scale_y_log10()

p2 <- ggplot() +
    geom_violin(data = dt, aes(arrayID, nCount_RNA, fill = protocol), scale = "width") +
    geom_point(data = dt %>% 
                 mutate(replicate = setNames(c("rep1", "rep2", "rep1", "rep1", "rep2", "rep1"), 
                                       nm = c("V11M22-349_A1", "V11M22-349_B1", "V11A20-396_A1",
                                              "V10S29-108_B1", "V11B18-363_C1", "V11A20-396_C1"))[arrayID]) %>%
                 group_by(labels, paper_id, arrayID, replicate) %>% 
                 summarize(Mean = median(nCount_RNA)), 
               aes(arrayID, Mean, shape = replicate), size = 3) +
    scale_shape_manual(values = c("rep1" = 4, "rep2" = 6)) +
    facet_grid(paper_id~labels, scales = "free") +
    labs(y = "", x = "", title = "UMI counts") +
    th +
    scale_y_log10()

p <- p1 | p2
p

```

```{r eval=FALSE}

# Export plot
pdf(file = "plots/violin_QC_by_region.pdf", width = 14*1.2, height = 6*1.2)
print(p)
dev.off()

```

```{r fig.width=9, fig.height=4}

CLN$unique_genes_log10 <- log10(CLN$nFeature_RNA)
CLN$UMIs_log10 <- log10(CLN$nCount_RNA)
CLN$UMIs_log10[CLN$UMIs_log10 < quantile(CLN$UMIs_log10, 0.01)] <- quantile(CLN$UMIs_log10, 0.01)
col_dims <- c(min(CLN$UMIs_log10), max(CLN$UMIs_log10))

ann_plots <- lapply(c("2", "3", "4", "6"), function(i) {
  gg <- cbind(CLN[[]], GetStaffli(CLN)@meta.data)
  gg <- subset(gg, sample == i)
  dims <- GetStaffli(CLN)@dims[[i]]
  
  p <- ggplot(gg, aes(warped_x, dims$height - warped_y, color = UMIs_log10)) +
    geom_point(size = 1) +
    theme_void() +
    scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "Spectral") %>% rev(), 
                          breaks = c(2, 3, 4), labels = c(100, 1000, 10000),
                          limits = col_dims) +
    scale_x_continuous(expand = c(0, 0), limits = c(limits_list[[i]]["x_start"], 
                                  limits_list[[i]]["x_end"])) +
    scale_y_continuous(expand = c(0, 0), limits = c(dims$height - limits_list[[i]]["y_end"], 
                                  dims$height - limits_list[[i]]["y_start"])) +
    theme(legend.position = "none")
  
  p <- ggrastr::rasterize(p, layers = "Point", dpi = 300)
  return(p)
})

ann_plots[[1]] + ann_plots[[2]]
ann_plots[[3]] + ann_plots[[4]]

```

```{r eval=FALSE}

pdf(file = "plots/sample1_RRST_UMIs_log10.pdf", width = 3.15, height = 3)
print(ann_plots[[1]])
dev.off()

pdf(file = "plots/sample1_standard_UMIs_log10.pdf", width = 2.75, height = 2.7)
print(ann_plots[[2]])
dev.off()

pdf(file = "plots/sample2_RRST_UMIs_log10.pdf", width = 3.9, height = 3)
print(ann_plots[[3]])
dev.off()

pdf(file = "plots/sample2_standard_UMIs_log10.pdf", width = 3.5, height = 3.4)
print(ann_plots[[4]])
dev.off()

```


Find genes that are detected in both RRST (RNA rescue = RRST) and standard data

```{r}

umis_standard <- rownames(CLN)[rowSums(GetAssayData(CLN, slot = "counts")[, CLN$protocol == "standard"]) > 0]
umis_RRST <- rownames(CLN)[rowSums(GetAssayData(CLN, slot = "counts")[, CLN$protocol == "RNA rescue"]) > 0]

genes.keep <- intersect(umis_standard, umis_RRST)

```

Subset and normalize data.

```{r}

CLN.subset <- SubsetSTData(CLN, features = genes.keep)
CLN.subset <- CLN.subset %>% 
  NormalizeData() %>% 
  ScaleData()

```


## Create plots for Figure 3 e
***

The marker genes were selected using the [Gut Cell Atlas](https://www.gutcellatlas.org/spacetime/full/)

Link to paper: https://doi.org/10.1038/s41586-021-03852-1

```{r fig.width=12, fig.height=2, out.width="100%"}

ST.FeaturePlot(CLN.subset, features = "labels", ncol = 6, show.sb = FALSE)

```
Spatial distribution of selected marker genes.

```{r fig.width=12, fig.height=22}

selected.genes <- c("CLDN3", "CLDN4", "CLDN7", "FABP1", "SLC26A2", "SLC26A3", "MS4A12", "CEACAM7", "AQP8", "KRT20", "PIGR")
ST.FeaturePlot(CLN.subset, features = selected.genes, ncol = 11, grid.ncol = 1, show.sb = F)

```

### Figure 3e
***

```{r fig.width=12, fig.height=5}

gg <- cbind(CLN.subset[[]], FetchData(CLN.subset, vars = selected.genes)) %>%
  subset(labels == "mucosa") %>%
  tidyr::gather(variable, value, selected.genes)

ggs <- gg %>% 
  group_by(protocol, paper_id, variable) %>%
  summarize(det = paste0(round(sum(value > 0)/n(), digits = 2)*100, "%"))
  

p <- ggplot() +
  geom_violin(data = gg, aes(protocol, value, fill = protocol), scale = "width") +
  geom_label(data = ggs, aes(protocol, -1, fill = protocol, label = det)) +
  facet_grid(paper_id ~ variable) +
  labs(y = "Normalized expression") +
  theme(axis.text.x = element_blank(), 
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 16, color = "black"),
        strip.text = element_text(size = 14, color = "black", face = "bold"),
        legend.text = element_text(size = 14),
        legend.title = element_text(size = 14),
        axis.text.y = element_text(size = 14, color = "black"),
        panel.background = element_rect(fill = "white"),
        panel.grid = element_line(colour = "lightgray", linetype = "longdash"),
        legend.position = "bottom") +
  scale_fill_brewer(palette = "Pastel1")
p <- ggrastr::rasterize(p, layers = "Point", dpi = 300)
p

```

```{r include=FALSE, eval=FALSE}
openxlsx::write.xlsx(ggs |> rename(gene = variable, detection_rate = det), file = "../../submission/source_data/source_data_3e.xlsx", sheetName = "3e")
```


```{r eval=FALSE}

pdf(file = "plots/selected_markers_violin.pdf", width = 18, height = 5.5)
print(p)
dev.off()

```


# date
***

```{r}
date()
```

# Session
***

```{r}
devtools::session_info()
```