---
title: "Figure 5"
author: "Ludvig Larsson"
date: '2022-08-17'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load libraries

```{r eval=FALSE}

library(STutility)
library(ggplot2)
library(ggpubr)
library(dplyr)
library(magrittr)

```

# Download files from Mendeley Data 
***

Note that you only have to run this code chunk once. If you are having issue downloading files, please make sure that 
you have a stable internet connection and that Mendeley Data (https://data.mendeley.com/) is up and running.

These files should be downloaded without any issue. If you encounter any errors, some of the code in this 
notebook will be possible to run.

If you struggle to download the data from R, you can also download the files manually from Mendeley Data and
put them in a folder called `data/` at the root of the project.

```{r eval=FALSE}

# Create data directories
suppressWarnings({
  dir.create("../data/")
  dir.create("../data/genes/")
  dir.create("../data/sheets/")
  dir.create("../data/sheets/PBT_pathology/")
  dir.create("../data/spaceranger_output/")
})

source("../scripts/mendeley.R")

# Download gene annotation files
for (i in seq_along(genes.files)) {
  download.file(url = genes.files[i], destfile = paste0("../data/genes/", names(genes.files)[i]))
}

# Download gene annotation files
for (i in seq_along(sheets.files)) {
  download.file(url = sheets.files[i], destfile = paste0("../data/sheets/", names(sheets.files)[i]))
}

# Download PBT pathology annotation files
for (i in seq_along(PBT.pathology.files)) {
  download.file(url = PBT.pathology.files[i], destfile = paste0("../data/sheets/PBT_pathology/", names(PBT.pathology.files)[i]))
}

# Download spaceranger output files
download.file(url = spaceranger.files[1], destfile = paste0("../data/spaceranger_output/", names(spaceranger.files)[1]))
untar(tarfile = paste0("../data/spaceranger_output/", names(spaceranger.files)[1]), exdir = "../data/spaceranger_output/")
file.remove(paste0("../data/spaceranger_output/", names(spaceranger.files)[1]))

```

Assemble spaceranger output files and merge curated meta data

```{r}

samples <- Sys.glob(paths = "../data/spaceranger_output/pediatricbraintumor/*/filtered_feature_bc_matrix.h5")
imgs <- Sys.glob(paths = "../data/spaceranger_output/pediatricbraintumor/*/spatial/tissue_hires_image.png")
spotfiles <- Sys.glob(paths = "../data/spaceranger_output/pediatricbraintumor/*/spatial/tissue_positions_list.csv")
json <- Sys.glob(paths = "../data/spaceranger_output/pediatricbraintumor/*/spatial/scalefactors_json.json")

infoTable <- data.frame(samples, imgs, spotfiles, json)
infoTable <- cbind(infoTable, arrayID = do.call(rbind, strsplit(infoTable$samples, "/"))[, 5])

curated_metadata <- openxlsx::read.xlsx("../data/sheets/RRST_sample_metadata.xlsx")
curated_metadata <- setNames(curated_metadata, nm = c("storage_time", "seq_date", "comments",
                                                       "ID", "paper_id", "RIN", "DV200", "protocol", "source", 
                                                       "arrayID", "spots_under_tissue",
                                                       "genes_detected", "fraction_spots_under_tissue",
                                                       "median_genes_per_spot", "median_UMIs_per_spot", 
                                                       "saturation", "reads_mapped_to_probe_set",
                                                       "reads_mapped_confidently_to_probe_set",
                                                       "reads_mapped_confidently_to_filtered_probe_set",
                                                       "reads_mapped_to_genome",
                                                       "reads_mapped_confidently_to_genome",
                                                       "number_of_panel_genes"))

infoTable <- merge(infoTable, curated_metadata, by = "arrayID")

```

Load data into a Seurat object

```{r}

CBT <- InputFromTable(infoTable)

```

```{r fig.width=12, fig.height=8}

CBT$protocol_array <- paste0(CBT$protocol, " : ", CBT$arrayID)
ST.FeaturePlot(CBT, features = "nFeature_RNA", ncol = 4, label.by = "protocol_array", show.sb = FALSE, pt.size = 1.5)

```

Load images

```{r}

CBT <- LoadImages(CBT, time.resolve = FALSE, xdim = 1e3)

```

```{r fig.width=8, fig.height=4}

ImagePlot(CBT, method = "raster", ncols = 4, type = "raw")

```

Apply rigid transformations to obtain a rough alignment of the tissue sections.

```{r}

# Warp transform
CBT <- WarpImages(CBT, verbose = TRUE, transforms = list("2" = list(shift.y = 40*2.5),
                                                         "3" = list(shift.y = 40*2.5),
                                                         "4" = list(shift.y = 40*2.5),
                                                         "6" = list(shift.y = 20*2.5),
                                                         "7" = list(shift.y = 30*2.5, shift.x = -30*2.5),
                                                         "8" = list(shift.x = 20*2.5)))

```

## Figure 5a - Violin plots of unique genes per spot
***

```{r fig.width=8, fig.height=6}

p_violin <- ggplot() + 
  geom_violin(data = CBT[[]], 
              aes(arrayID, nFeature_RNA, fill = protocol), scale = "width") + 
  geom_hline(data = CBT[[]] %>% group_by(protocol, paper_id) %>%
               summarize(Mean = round(mean(nFeature_RNA))), 
             aes(yintercept = Mean), linetype = "longdash") +
  geom_label(data = CBT[[]] %>% group_by(protocol, paper_id) %>%
               summarize(Mean = round(mean(nFeature_RNA))), 
             aes("A", Mean, label = Mean), size = 7) +
  facet_grid(~paper_id, scales = "free", space = "free") +
  scale_x_discrete(labels = c("", "Rep1", "Rep2", "Rep1", "Rep2")) +
  scale_y_log10() +
  labs(x = "", y = "") +
  theme(panel.background = element_rect(fill = "white"), panel.grid = element_line(colour = "lightgray", size = 0.5), 
        strip.text = element_text(size = 16, colour = "black"), 
        axis.text = element_text(size = 16, colour = "black"), 
        legend.text = element_text(size = 14, colour = "black"),
        legend.title = element_text(size = 16, colour = "black"), 
        legend.position = "bottom")
p_violin

```

```{r eval=FALSE}

# Export plot
pdf(file = "plots/unique_genes_overview.pdf", width = 8, height = 6)
print(p_violin)
dev.off()

```

## Figure 5b - Spatial maps of unique genes per spot and H&E images
***

```{r}

# Spatial plots showing the number of unique genes
df_subset <- cbind(CBT[[]], GetStaffli(CBT)@meta.data) %>%
  subset(sample %in% c("1", "3", "5", "7"))
scale_limits <- c(0, max(df_subset$nFeature_RNA))
x_limits <- range(df_subset$warped_x)
y_limits <- range(df_subset$warped_y)
p_spatial <- df_subset %>%
    ggplot(aes(warped_x, 2000 - warped_y, color = nFeature_RNA)) +
      geom_point(size = 1) +
      facet_wrap(paper_id ~ protocol, labeller = labeller(paper_id = c("1.0" = "CBT1", "2.0" = "CBT2")), ncol = 4) +
      theme_void() +
      labs(color = "Unique\ngenes") +
      scale_color_gradientn(colours = viridis::magma(n = 11, direction = -1), limits = scale_limits) +
      scale_x_continuous(limits = x_limits, expand = c(0, 0)) +
      scale_y_continuous(limits = c(2000 - y_limits[2], 2000 - y_limits[1]), expand = c(0, 0)) +
      theme(strip.text = element_blank(), 
            panel.background = element_rect(colour = "lightgray", size = 2),
            legend.text = element_text(size = 14, colour = "black"),
            legend.title = element_text(size = 16, colour = "black"))
p_spatial <- ggrastr::rasterize(p_spatial, layers = "Point", dpi = 300)

# HE images
plots_HE <- lapply(c("1", "3", "5", "7"), function(i) {
  im <- GetStaffli(CBT)@rasterlists$processed[[i]]
   
  im <- im[round(y_limits[1]*0.5):round( y_limits[2]*0.5), round(x_limits[1]*0.5):round(x_limits[2]*0.5)]
  g <- grid::rasterGrob(im, width = unit(1, "npc"), height = unit(1, "npc"), interpolate = TRUE)
  p <- ggplot() +
    annotation_custom(g, -Inf, Inf, -Inf, Inf) +
    theme_void() +
    theme(plot.margin = margin(t = 1, r = 5, b = 1, l = 5))
  return(p)
})

```

Create patchwork for Figure 5b

```{r fig.width=12, fig.height=5}

design <- c(
  patchwork::area(1, 1, 1, 1),
  patchwork::area(1, 2, 1, 2),
  patchwork::area(1, 3, 1, 3),
  patchwork::area(1, 4, 1, 4),
  patchwork::area(2, 1, 2, 4)
)

# Create patchwork
p_HE <- plots_HE[[3]] +
  plots_HE[[1]] +
  plots_HE[[4]] +
  plots_HE[[2]] +
  p_spatial +
  patchwork::plot_layout(design = design)

p_HE

```

```{r eval=FALSE}

# Export plot
pdf(file = "plots/HE_overivew.pdf", width = 14, height = 6)
print(p_HE)
dev.off()

```

Import annotations made by the pathologist

```{r}

annotations <- Sys.glob(paths = "../data/sheets/PBT_pathology/*")

ann <- do.call(rbind, lapply(seq_along(annotations), function(i) {
  a <- setNames(read.csv(file = annotations[i]), nm = c("barcode", "type"))
  a <- data.frame(a, row.names = 1)
  rownames(a) <- paste0(rownames(a), "_", i + 4)
  return(a)
}))

CBT$annotation <- ann[colnames(CBT), "type"]
CBT$annotation[CBT$annotation == "N/A"] <- "unassigned"

```

## Suppl. figure - PBT annotations
***

```{r fig.width=6, fig.height=5}

p <- FeatureOverlay(CBT, features = "annotation", sampleids = 5, show.sb = FALSE) &
  theme(plot.title = element_blank(), plot.subtitle = element_blank())

im <- GetStaffli(CBT)@rasterlists$processed[[5]]
im <- im[round(500*0.5):round( y_limits[2]*0.5), round(x_limits[1]*0.5):round(x_limits[2]*0.5)]
g <- grid::rasterGrob(im, width = unit(1, "npc"), height = unit(1, "npc"), interpolate = TRUE)
p <- ggplot() +
  annotation_custom(g, -Inf, Inf, -Inf, Inf) +
  geom_point(data = cbind(CBT[[]], GetStaffli(CBT)@meta.data) %>% subset(sample == "5"), 
             aes(warped_x, 2000 - warped_y, color = annotation), alpha = 1) +
  scale_x_continuous(limits = x_limits, expand = c(0, 0)) +
  scale_y_continuous(limits = c(2000 - y_limits[2], 2000 - 500), expand = c(0, 0)) +
  scale_color_manual(values = c("tumor" = "#117733", "stroma" = "#DDCC77", "unassigned" = "lightgrey")) +
  theme_void() +
  guides(colour = guide_legend(override.aes = list(size = 5))) +
  theme(plot.margin = margin(t = 1, r = 5, b = 1, l = 5), 
        legend.position = "bottom", 
        legend.text = element_text(size = 14, colour = "black"),
        legend.title = element_blank())
p <- ggrastr::rasterize(p, layers = "Point", dpi = 300)

p

```

```{r eval=FALSE}

pdf(file = "../Suppl_figures/Suppl_Figure_PBT_anntoations/medulloblastoma_annotations.pdf", width = 5, height = 4.2)
print(p)
dev.off()


```


Subset data to include medulloblastoma only

```{r fig.width=8, fig.height=3}

CBT_medulloblastoma <- SubsetSTData(CBT, expression = paper_id == "medulloblastoma")
CBT_medulloblastoma <- CBT_medulloblastoma %>%
  NormalizeData() %>%
  ScaleData() %>%
  FindVariableFeatures()
ST.FeaturePlot(CBT_medulloblastoma, features = "AXIN2", ncol = 4, show.sb = FALSE)

```

## Figure 5c - Violin plots of marker gene expression
***

Visualize the normalized expression of marker genes associated with WNT-signaling.

```{r fig.width=12, fig.height=7}

WNT_markers <- c("SP5", "AXIN2", "LEF1", "DKK4", "GAD1", "CTNNB1")

p_markers <- cbind(CBT_medulloblastoma[[]], FetchData(CBT_medulloblastoma, vars = WNT_markers)) %>%
  reshape2::melt(measure.vars = WNT_markers) %>%
  ggplot(aes(arrayID, value, fill = protocol)) +
    geom_violin(scale = "width") +
    scale_x_discrete(labels = c("Rep1", "Rep2", "Rep1", "Rep2")) +
    facet_wrap(~variable, ncol = 3) +
    theme(panel.background = element_rect(fill = "white"),
          panel.grid = element_line(linetype = "longdash", colour = "lightgrey"),
          strip.text = element_text(size = 14, colour = "black"),
          axis.text.x = element_text(size = 12, colour = "black"),
          axis.text.y = element_text(size = 14, colour = "black"),
          axis.title = element_text(size = 16, colour = "black")) +
  labs(y = "Norm. Expr.", x = "")
p_markers

```

```{r eval=FALSE}

pdf(file = "plots/violin_markers.pdf", width = 8, height = 5)
print(p_markers)
dev.off()

```


Calculate module score for WNT signaling (hallmark WNT beta catenin signaling).

```{r}

# Function sued to read gmt files
read.gmt <- function(filename) {
  d <- suppressWarnings({(readLines(filename) %>% strsplit("\t"))[[1]]})
  setNames(list(d[3:length(d)]), nm = d[1])
}

# readt gmt file containing WNT gene set
WNT_genes <- read.gmt(filename = "pathway_gmt_files/HALLMARK_WNT_BETA_CATENIN_SIGNALING.v7.5.1.gmt")

# Compute module score
CBT_medulloblastoma <- AddModuleScore(CBT_medulloblastoma, features = WNT_genes, name = "pathway")

```

## Figure 5d - Spatial maps of WNT module scores
***

```{r fig.width=8, fig.height=5}

gg <- cbind(CBT_medulloblastoma[[]], GetStaffli(CBT_medulloblastoma)@meta.data) %>%
  mutate(pathway1 = ifelse(pathway1 > quantile(pathway1, 0.99), quantile(pathway1, 0.99), pathway1))

p_pathway <- gg %>%
  ggplot(aes(warped_x, 2000 - warped_y, color = pathway1)) +
    geom_point(size = 1) +
    facet_wrap(arrayID ~ protocol, dir = "v") +
    scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev(), 
                          limits = c(-max(abs(gg$pathway1)), max(abs(gg$pathway1)))) +
    theme_void() +
    theme(strip.text = element_blank(), panel.background = element_rect(colour = "lightgrey", size = 2),
          legend.text = element_text(size = 14, colour = "black"),
          legend.title = element_text(size = 16, colour = "black"))
p_pathway <- ggrastr::rasterize(p_pathway, layers = "Point", dpi = 300)
p_pathway

```

```{r eval=FALSE}

pdf(file = "plots/pathway_spatial_plot.pdf", width = 9, height = 6.5)
print(p_pathway)
dev.off()

```

# date
***

```{r}

date()

```

# Session
***

```{r}

sessionInfo()

```

