---
title: "Figure 2 + suppl.figures"
author: "Ludvig Larsson"
date: '2022-05-23'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load libraries

```{r warning=FALSE, message=FALSE}

library(STutility)
library(ggplot2)
library(ggpubr)
library(dplyr)
library(magrittr)
library(scico)

```

# Download files from Mendeley Data 
***

Note that you only have to run this code chunk once. If you are having issue downloading files, please make sure that 
you have a stable internet connection and that Mendeley Data (https://data.mendeley.com/) is up and running.

These files should be downloaded without any issue. If you encounter any errors, some of the code in this 
notebook will be possible to run.

If you struggle to download the data from R, you can also download the files manually from Mendeley Data and
put them in a folder called `data/` at the root of the project.

```{r eval=FALSE}

# Create data directories
suppressWarnings({
  dir.create("../data/")
  dir.create("../data/genes/")
  dir.create("../data/sheets/")
  dir.create("../data/spaceranger_output/")
})

source("../scripts/mendeley.R")

# Download gene annotation files
for (i in seq_along(genes.files)) {
  download.file(url = genes.files[i], destfile = paste0("../data/genes/", names(genes.files)[i]))
}

# Download gene annotation files
for (i in seq_along(sheets.files)) {
  download.file(url = sheets.files[i], destfile = paste0("../data/sheets/", names(sheets.files)[i]))
}

# Download spaceranger output files
download.file(url = spaceranger.files[3], destfile = paste0("../data/spaceranger_output/", names(spaceranger.files)[3]))
untar(tarfile = paste0("../data/spaceranger_output/", names(spaceranger.files)[3]), exdir = "../data/spaceranger_output/")
file.remove(paste0("../data/spaceranger_output/", names(spaceranger.files)[3]))

```

Assemble spaceranger output files and merge curated meta data

```{r}

samples <- Sys.glob(paths = "../data/spaceranger_output/lung/*/filtered_feature_bc_matrix.h5")
imgs <- Sys.glob(paths = "../data/spaceranger_output/lung/*/spatial/tissue_hires_image.png")
spotfiles <- Sys.glob(paths = "../data/spaceranger_output/lung/*/spatial/tissue_positions_list.csv")
json <- Sys.glob(paths = "../data/spaceranger_output/lung/*/spatial/scalefactors_json.json")

infoTable <- data.frame(samples, imgs, spotfiles, json)
infoTable <- cbind(infoTable, arrayID = do.call(rbind, strsplit(infoTable$samples, "/"))[, 5])

curated_metadata <- openxlsx::read.xlsx("../data/sheets/RRST_sample_metadata.xlsx")
curated_metadata <- setNames(curated_metadata, nm = c("storage_time", "seq_date", "comments",
                                                       "ID", "paper_id", "RIN", "DV200", "protocol", "source", 
                                                       "arrayID", "spots_under_tissue",
                                                       "genes_detected", "fraction_spots_under_tissue",
                                                       "median_genes_per_spot", "median_UMIs_per_spot", 
                                                       "saturation", "reads_mapped_to_probe_set",
                                                       "reads_mapped_confidently_to_probe_set",
                                                       "reads_mapped_confidently_to_filtered_probe_set",
                                                       "reads_mapped_to_genome",
                                                       "reads_mapped_confidently_to_genome",
                                                       "number_of_panel_genes"))

infoTable <- merge(infoTable, curated_metadata, by = "arrayID")

```

Load data into a Seurat object

* 1 : LNG1, RRST
* 2-3 : LNG1, standard
* 4 : LNG2, RRST 
* 5-6 : LNG2, standard

```{r}

LNG <- InputFromTable(infoTable[c(1, 3:4, 2, 5:6), ])

```

Load images

```{r}

LNG <- LoadImages(LNG, time.resolve = FALSE, xdim = 1e3)

```

Apply rigid transformations, e.g. rotations to get a rough alignment of the H&E images

```{r}

# Warp transform
LNG <- WarpImages(LNG, verbose = TRUE, transforms = list( "2" = list(angle = -90), 
                                                          "3" = list(angle = -90), 
                                                          "5" = list(angle = 180)))

```

Add a new metadata column with a combined protocol and lung ID label

```{r fig.width=12, fig.height=9}

LNG$protocol_lng <- gsub(pattern = " ", replacement = "_", paste0(LNG$protocol, "_ID: ", LNG$paper_id))
ST.FeaturePlot(LNG, features = "nFeature_RNA", ncol = 3, label.by = "protocol_lng")

```

## QC
***

The standard Visium data sets have reduced numbers of unique genes compared to the data generated by RRST.

```{r}

VlnPlot(LNG, features = "nFeature_RNA", group.by = "arrayID", split.by = "protocol")

```

At a first glance, we can see that the distribution of unique genes is quite different across the two protocols. For RRST, we have much more even coverage across the tissue, whereas in the standard Visium data, we only get high counts in the epithelium. 

A reasonable threshold for filtering would be somewhere between 300-1000 unique genes, depending on the tissue type. If parts of the tissue is cell sparse, which is the case for the alveoli, it might be more reasonable to set a lower threshold. 

If we use a threshold of 300 unique genes, we can see that ~1.3%-2.3% of spots are discarded from the RR-seq data, whereas ~21%-80% of spots are discarded from the standard Visium data.

## Suppl. Figure - histograms (filters) for lung Visium data
***

```{r}

LNG[[]] %>% 
  dplyr::mutate(filter = ifelse(nFeature_RNA < 300, "discard", "keep")) %>%
  dplyr::group_by(protocol, paper_id, filter) %>%
  dplyr::summarize(Freq = n()) %>%
  dplyr::group_by(protocol, paper_id) %>%
  dplyr::mutate(percentage_discarded = round(Freq/sum(Freq)*100, digits = 2)) %>%
  dplyr::filter(filter == "discard") %>%
  dplyr::select(-Freq, -filter) %>%
  dplyr::arrange(paper_id)

p <- LNG[[]] %>% ggplot(aes(nFeature_RNA, fill = ifelse(nFeature_RNA < 300, "discard", "keep"))) +
  geom_histogram(binwidth = 100) +
  geom_vline(xintercept = 300, linetype = "longdash") +
  facet_grid(protocol~paper_id, scales = "free") +
  labs(fill = "", x = "Unique genes", y = "Count") +
  scale_x_continuous(breaks = c(0, 300, 2000, 4000, 6000), minor_breaks = NULL) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1), 
        axis.text = element_text(size = 10, colour = "black"),
        axis.title = element_text(size = 14, colour = "black"),
        strip.text = element_text(size = 14, colour = "black", face = "bold"),
        legend.text = element_text(size = 10, colour = "black"))

p

```

```{r eval=FALSE}

pdf(file = "../Suppl_figures/Suppl_Figure_filter_lung_data/Suppl_Figure_2.pdf", width = 10, height = 7)
print(p)
dev.off()

png(filename = "../Suppl_figures/Suppl_Figure_filter_lung_data/Suppl_Figure_2.png", width = 2400, height = 1600, res = 300)
print(p)
dev.off()

```

### Filter data
***

```{r}

LNG$filter <- ifelse(LNG$nFeature_RNA < 300, "discard", "keep")
LNG.subset <- SubsetSTData(LNG, expression = nFeature_RNA > 300)

```

Now we can see that quite a few spots have been discarded, in particular for the standard Visium data for LNG2.

```{r fig.width=12, fig.height=9}

ST.FeaturePlot(LNG.subset, features = "nFeature_RNA", ncol = 3, label.by = "protocol_lng") &
  theme(panel.background = element_rect(fill = "lightgrey"))

```

```{r eval=FALSE, include=FALSE}

df_qc_LNG1 <- LNG[[]] |> 
  select(nFeature_RNA, nCount_RNA, paper_id, filter, protocol) |> 
  rename(unique_genes = nFeature_RNA, UMIs = nCount_RNA) |> 
  bind_cols(GetStaffli(LNG)@meta.data |> select(sample, warped_x, warped_y)) |> 
  filter(paper_id %in% "LNG1")

openxlsx::write.xlsx(df_qc_LNG1 |> select(unique_genes, UMIs, protocol), file = "../../submission/source_data/source_data_2b.xlsx", sheetName = "2b")

openxlsx::write.xlsx(df_qc_LNG1 |> select(unique_genes, warped_x, warped_y), file = "../../submission/source_data/source_data_2c.xlsx", sheetName = "2x")

openxlsx::write.xlsx(df_qc_LNG1 |> select(filter, warped_x, warped_y), file = "../../submission/source_data/source_data_2d.xlsx", sheetName = "2d")

```


Split data by protocol and sample

```{r eval=FALSE}

LNGList <- lapply(unique(LNG.subset$protocol_lng), function(i) {
  SubsetSTData(LNG.subset, expression = protocol_lng == i)
})

```

## Analysis
***

1. log-normalize data
2. scale data (z-transformation)
3. find top variable features
4. run dimensionality reduction using Principal Component Analysis (PCA)
5. Create an SNN graph using the first 20 PCs as input
6. Find clusters using Louvain algorithm
7. Create a UMAP using the first 20 PCs as input


```{r eval=FALSE}

LNGList <- lapply(LNGList, function(LNG.subset) {
  LNG.subset <- LNG.subset %>%
    NormalizeData() %>%
    ScaleData() %>%
    FindVariableFeatures() %>%
    RunPCA() %>%
    FindNeighbors(reduction = "pca", dims = 1:20) %>%
    FindClusters(resolution = 0.8) %>%
    RunUMAP(reduction = "pca", dims = 1:20, min.dist = 0.3, n.epochs = 1e3)
})
LNGList <- setNames(LNGList, nm = unique(LNG$protocol_lng))

```
```{r include=FALSE, eval=FALSE}

saveRDS(LNGList, file = "R_objects/LNGList")

```

Plot clusters in UMAP embedding

```{r fig.width=9, fig.height=4}

DimPlot(LNGList[[1]]) & ggtitle("LNG1 - RRST") | DimPlot(LNGList[[2]]) & ggtitle("LNG1 - standard")
DimPlot(LNGList[[3]]) & ggtitle("LNG2 - RRST") | DimPlot(LNGList[[4]]) & ggtitle("LNG2 - standard")

```

```{r eval=FALSE, include=FALSE}

df_clusters <- LNGList[[1]][[]] |> 
  bind_cols(GetStaffli(LNGList[[1]])@meta.data |> select(warped_x, warped_y)) |> 
  bind_cols(FetchData(LNGList[[1]], vars = c("UMAP_1","UMAP_2"))) |> 
  bind_rows(LNGList[[2]][[]] |> 
              bind_cols(GetStaffli(LNGList[[2]])@meta.data |> select(warped_x, warped_y)) |> 
              bind_cols(FetchData(LNGList[[2]], vars = c("UMAP_1","UMAP_2"))))

openxlsx::write.xlsx(df_clusters |> select(UMAP_1, UMAP_2, protocol, seurat_clusters), 
                     file = "../../submission/source_data/source_data_2e.xlsx", sheetName = "2e")

openxlsx::write.xlsx(df_clusters |> select(warped_x, warped_y, protocol, seurat_clusters), 
                     file = "../../submission/source_data/source_data_2f.xlsx", sheetName = "2f")

```


## DEA
***

Run DEA with an avg_log2FC threshold of 0.5 and adjusted p-value < 0.01.

```{r eval=FALSE}

# Keep lower threshold avg_log2FC for checking DE genes in all clusters
de.markers.list.025 <- lapply(LNGList, function(se) {
  se <- SetIdent(se, value = "seurat_clusters")
  FindAllMarkers(se, only.pos = TRUE, logfc.threshold = 0.25)
})
de.markers.list.025 <- lapply(de.markers.list.025, function(de.markers) {
  de.markers %>% dplyr::filter(p_val_adj < 0.01)
})

# avg_log2FC > 0.5 
# These are the DE genes used for figure plots
de.markers.list <- lapply(de.markers.list.025, function(de.markers) {
  de.markers %>% dplyr::filter(avg_log2FC > 0.5)
})


```

```{r eval=FALSE, include=FALSE}

# only run if you want to export DE list objects
saveRDS(de.markers.list.025, file = "R_objects/de.markers.list.025")
saveRDS(de.markers.list, file = "R_objects/de.markers.list")

```


### Add cluster annotations
***

```{r}

# RRST cluster annotations
ann1 <- c("0" = "AT1-enriched", "1" = "Macrophage-enriched",
         "2" = "AT2-enriched", "3" = "Connective tissue 1", 
         "4" = "Connective tissue 2", "5" = "Connective tissue 3",
         "6" = "Airway epithelium", "7" = "Megakaryocyte/platelet-enriched",
         "8" = "Smooth muscle", "9" = "Immune-enriched", 
         "10" = "Glands")
ann3 <- c("0" = "AT2-enriched", "1" = "Low-quality cluster 1",
          "2" = "Connective tissue 1", "3" = "Connective tissue 2",
          "4" = "Low-quality cluster 2", "5" = "Smooth muscle",
          "6" = "Glands", "7" = "Megakaryocyte/platelet-enriched", 
          "8" = "Airway epithelium")

```


### Create plots for figure 2
***

* plots_vln : violin plots showing the number of unique genes and UMIs 
* plots_filter : spatial plots showing what spots are discarded in the filtering step (unique genes < 300)
* plots_umap : UMAP plots colored by cluster
* plots_qc : spatial plots showing the number of unique genes
* plots_HE : HE images
* plots_spatial : spatial plots showing clusters, split by panel
* plots_dotplot : dot plot showing DE genes per cluster

```{r}

# Violin plots showing the number of unique genes and UMIs
plots_vln <- lapply(1:4, function(i) {
  p <- LNGList[[i]][[]] %>%
    dplyr::mutate(`Unique genes` = nFeature_RNA, `UMI counts` = nCount_RNA) %>%
    reshape2::melt(measure.vars = c("Unique genes", "UMI counts")) %>%
    ggplot(aes(paper_id, value, fill = variable)) +
    geom_violin(scale = "width", alpha = 0.7) +
    geom_boxplot(width = 0.3) +
    scale_y_log10() +
    theme_classic() +
    facet_wrap(~variable) +
    theme(axis.title = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
          legend.position = "none", axis.text.y = element_text(size = 18, colour = "black"), 
          strip.text = element_text(size = 18))
  return(p)
})

# Spatial plots showing the spots discarded
plots_filter <- lapply(c("RRST_ID:_LNG1", "standard_ID:_LNG1", "RRST_ID:_LNG1", "standard_ID:_LNG1"), function(i) {
  p <- ST.FeaturePlot(SubsetSTData(LNG, expression = protocol_lng %in% i), features = "filter", show.sb = FALSE, pt.size = 1.3, 
                 cols = c("keep" = "lightgray", "discard" = "red"), 
                 indices = ifelse(i %in% c("standard_ID:_LNG1", "standard_ID:_LNG2"), 2, 1)) &
    theme(strip.text = element_blank(), plot.title = element_blank(), legend.title = element_blank(), legend.text = element_text(size = 16)) &
  guides(fill = guide_legend(override.aes = list(size = 7)))
  p <- ggrastr::rasterize(p, layers = "Point", dpi = 300)
  return(p)
})

# UMAP plots showing clusters
plots_umap <- lapply(1:4, function(i) {
  gg <- cbind(LNGList[[i]][[]], LNGList[[i]]@reductions$umap@cell.embeddings)

  p <- ggplot() +
    geom_point(data = gg, aes(UMAP_1, UMAP_2, color = seurat_clusters)) +
    ggrepel::geom_label_repel(size = 9, label.size = 0.2, data = gg %>% 
                 dplyr::group_by(seurat_clusters) %>% 
                 summarize(med1 = median(UMAP_1), med2 = median(UMAP_2)), aes(med1, med2 + 0.6, fill = seurat_clusters, label = seurat_clusters)) +
    theme_classic() +
    theme(plot.title = element_blank(), 
          legend.position = "none", axis.text = element_text(size = 14, colour = "black"), axis.title = element_text(size = 18))
  p <- ggrastr::rasterize(p, layers = "Point", dpi = 300)
  return(p)
})

# Spatial plots showing the number of unique genes
plots_qc <- lapply(1:4, function(i) {
  p <- ST.FeaturePlot(LNGList[[i]], features = "nFeature_RNA", show.sb = FALSE, pt.size = 1.3, 
                      indices = ifelse(i %in% c(2, 4), 2, 1), cols = viridis::magma(n = 11, direction = -1)) &
    theme(plot.subtitle = element_blank(), 
          plot.title = element_blank(), 
          strip.text = element_blank(), 
          legend.text = element_text(size = 14), 
          legend.title = element_text(size = 18)) &
    labs(fill = "Unique\ngenes") 
  p <- ggrastr::rasterize(p, layers = "Point", dpi = 300)
  return(p)
})

# HE images
plots_HE <- lapply(1:4, function(i) {
  im <- GetStaffli(LNGList[[i]])@rasterlists$processed[[ifelse(i %in% c(2, 4), 2, 1)]]
  im[, 1:100] <- "#ffffffff"; im[, (ncol(im) - 120):ncol(im)] <- "#ffffffff"
  im <- im[101:(nrow(im) - 100), ]
  g <- grid::rasterGrob(im, width = unit(1, "npc"), height = unit(1, "npc"), interpolate = TRUE)
  ggplot() +
    annotation_custom(g, -Inf, Inf, -Inf, Inf) +
    theme_void() +
    theme(plot.margin = margin(t = 0, r = 40, b = 0, l = 40))
})

# Spatial plots showing clusters split by clusters
plots_spatial <- lapply(1:4, function(i) {
  p <- ST.FeaturePlot(LNGList[[i]], features = "seurat_clusters", split.labels = TRUE, ncol = ifelse(i %in% c(1, 2), 4, 3), 
                 indices = ifelse(i %in% c(2, 4), 2, 1), show.sb = FALSE, pt.size = 0.5) +
    theme(plot.title = element_blank(), legend.position = "none", strip.text = element_blank()) &
    scale_x_continuous(limits = c(150, 1850)) &
    scale_y_continuous(limits = c(150, 1850))
  p <- ggrastr::rasterize(p, layers = "Point", dpi = 300)
  return(p)
})

# Dot plot showing DE genes
plots_dotplot <- lapply(1:4, function(i) {
  LNG <- LNGList[[i]]
  if (i == 1) {
    LNG$seurat_clusters <- factor(ann1[LNG$seurat_clusters], levels = ann1)
  }
  if (i == 2) {
    LNG$seurat_clusters <- factor(ann3[LNG$seurat_clusters], levels = ann3)
  }
  LNG <- Seurat::SetIdent(LNG, value = "seurat_clusters")
  p <- DotPlot(LNG, features = unique(de.markers.list[[i]] %>% 
                                          dplyr::mutate(gene_count = table(gene)[gene]) %>%
                                          dplyr::filter(gene_count == 1) %>%
                                          dplyr::group_by(cluster) %>%
                                          dplyr::slice_head(n = 4) %>% 
                                          dplyr::pull(gene) %>% rev())) +
    scale_color_gradientn(colours = scico::scico(n = 11, palette = "bam", direction = -1)) +
    scale_y_discrete(position = "left") +
    theme(axis.title = element_blank(), axis.text = element_text(size = 14), legend.text = element_text(size = 14), 
          legend.title = element_text(size = 14), axis.text.x = element_text(angle = 60, hjust = 1, size = 16)) +
    coord_flip()
  p$guides$colour$title <- "Avg. Expr."
  p$guides$size$title <- "Pct. Expr."
  return(p)
})

```

```{r eval=FALSE, include=FALSE}
degs <- do.call(rbind, lapply(c(1, 2), function(i) {
  LNG <- LNGList[[i]]
  if (i == 1) {
    LNG$seurat_clusters <- factor(ann1[LNG$seurat_clusters], levels = ann1)
  }
  if (i == 2) {
    LNG$seurat_clusters <- factor(ann3[LNG$seurat_clusters], levels = ann3)
  }
  LNG <- Seurat::SetIdent(LNG, value = "seurat_clusters")
  p <- DotPlot(LNG, features = unique(de.markers.list[[i]] %>% 
                                          dplyr::mutate(gene_count = table(gene)[gene]) %>%
                                          dplyr::filter(gene_count == 1) %>%
                                          dplyr::group_by(cluster) %>%
                                          dplyr::slice_head(n = 4) %>% 
                                          dplyr::pull(gene) %>% rev())) +
    scale_color_gradientn(colours = scico::scico(n = 11, palette = "bam", direction = -1)) +
    scale_y_discrete(position = "left") +
    theme(axis.title = element_blank(), axis.text = element_text(size = 14), legend.text = element_text(size = 14), 
          legend.title = element_text(size = 14), axis.text.x = element_text(angle = 60, hjust = 1, size = 16)) +
    coord_flip()
  p$guides$colour$title <- "Avg. Expr."
  p$guides$size$title <- "Pct. Expr."
  d <- p$data |> setNames(nm = c("V1", "Pct. Expr.", "gene", "annotation", "Avg. Expr.")) |> 
                       select(-V1)
  if (i == 1) {
    d$protocol <- "RRST"
  }
  if (i == 2) {
    d$protocol <- "standard"
  }
  return(d)
}))

openxlsx::write.xlsx(degs, file = "../../submission/source_data/source_data_2g.xlsx", sheetName = "2g")
```


### Patchwork for Figure 2 
***

This patchwork includes only LNG1 and each plot type is places side by side for a simple comparison between the two protocols.

```{r fig.width=26, fig.height=18}

# Design for patchwork
design <- c(
  patchwork::area(1, 1, 1, 1),
  patchwork::area(1, 2, 1, 2),
  patchwork::area(2, 1, 2, 1),
  patchwork::area(2, 2, 2, 2),
  patchwork::area(3, 1, 3, 1),
  patchwork::area(4, 1, 4, 1),
  patchwork::area(3, 2, 3, 2),
  patchwork::area(4, 2, 4, 2),
  
  patchwork::area(1, 3, 1, 3),
  patchwork::area(1, 4, 1, 4),
  patchwork::area(2, 3, 2, 3),
  patchwork::area(2, 4, 2, 4),
  patchwork::area(3, 3, 5, 3),
  patchwork::area(3, 4, 5, 4)
)

# Create patchwork
p <- plots_HE[[1]] +
  plots_HE[[2]] +
  plots_qc[[1]] +
  plots_qc[[2]] +
  plots_umap[[1]] + 
  plots_spatial[[1]] +
  plots_umap[[2]] +
  plots_spatial[[2]] +
  plots_vln[[1]] +
  plots_vln[[2]] +
  plots_filter[[1]] +
  plots_filter[[2]] +
  plots_dotplot[[1]] +
  plots_dotplot[[2]] +
  patchwork::plot_layout(design = design, heights = c(1.5, 2, 2, 2, 0.5))

p

```

```{r eval=FALSE}

# Export patchwork as pdf
pdf(file = "plots/figure_2_patchwork_side_by_side.pdf", width = 26, height = 20)
print(p)
dev.off()

```

### Patchwork for Supplementary Figure - LNG2 analysis results
***

This patchwork includes only LNG2 and each plot type is places side by side for a simple comparison between the two protocols.

```{r fig.width=26, fig.height=18}

# Design for patchwork
design <- c(
  patchwork::area(1, 1, 1, 1),
  patchwork::area(1, 2, 1, 2),
  patchwork::area(2, 1, 2, 1),
  patchwork::area(2, 2, 2, 2),
  patchwork::area(3, 1, 3, 1),
  patchwork::area(4, 1, 4, 1),
  patchwork::area(3, 2, 3, 2),
  patchwork::area(4, 2, 4, 2),
  
  patchwork::area(1, 3, 1, 3),
  patchwork::area(1, 4, 1, 4),
  patchwork::area(2, 3, 2, 3),
  patchwork::area(2, 4, 2, 4),
  patchwork::area(3, 3, 5, 3),
  patchwork::area(3, 4, 5, 4)
)

# Create patchwork
p <- plots_HE[[3]] +
  plots_HE[[4]] +
  plots_qc[[3]] +
  plots_qc[[4]] +
  plots_umap[[3]] + 
  plots_spatial[[3]] +
  plots_umap[[4]] +
  plots_spatial[[4]] +
  plots_vln[[3]] +
  plots_vln[[4]] +
  plots_filter[[3]] +
  plots_filter[[4]] +
  plots_dotplot[[3]] +
  plots_dotplot[[4]] +
  patchwork::plot_layout(design = design, heights = c(1.5, 2, 2, 2, 0.5))

p

```

```{r eval=FALSE}

# Export image
pdf(file = "../Suppl_figures/Suppl_Figure_lung_LNG2_analysis/suppl_figure_lung_LNG2_patchwork.pdf", width = 26, height = 20)
print(p)
dev.off()

```


### Suppl. figure - lung selected markers
***

Here we DEGs for clusters annotated as "Airway epithelium", "Glands", "Smooth muscle" and "Megakaryocyte/platelet-enriched" that are shared across conditions. 

```{r fig.width=2, fig.height=4}

selected_annotations <- c("Airway epithelium", "Glands", "Smooth muscle", "Megakaryocyte/platelet-enriched")
shared_markers <- de.markers.list.025$`RRST_ID:_LNG1` |> 
  mutate(annotation = ann1[cluster], protocol = "RRST") |> 
  bind_rows(de.markers.list.025$`standard_ID:_LNG1` |> mutate(annotation = ann3[cluster], protocol = "standard")) |> 
  filter(annotation %in% selected_annotations) |> 
  group_by(annotation, gene) |> 
  mutate(n = n())  |> 
  filter(n == 2) |> # Make sure that genes are detected with both protocols
  group_by(annotation, gene) |> 
  summarize(shared_avg_log2FC = mean(avg_log2FC), pct = paste0(pct.1, collapse = "_")) |> 
  arrange(annotation, -shared_avg_log2FC) |> 
  mutate(annotation = factor(annotation, selected_annotations)) |> 
  group_by(annotation) |> 
  slice_head(n = 100) |> # Keep maximum 100 genes
  separate(col = pct, sep = "_", into = c("pct_RRST", "pct_standard")) |> 
  mutate(across(pct_RRST:pct_standard, ~ as.numeric(.x)))

# Subset data to include selected clusters
LNG1_RRST <- LNGList[[1]]
LNG1_RRST$annotation <- ann1[LNG1_RRST$seurat_clusters |> as.character()]
LNG1_RRST <- SubsetSTData(LNG1_RRST, expression = annotation %in% selected_annotations)
LNG1_RRST$annotation <- factor(LNG1_RRST$annotation, levels = selected_annotations)
LNG1_standard <- LNGList[[2]]
LNG1_standard$annotation <- ann3[LNG1_standard$seurat_clusters |> as.character()]
LNG1_standard <- SubsetSTData(LNG1_standard, expression = annotation %in% selected_annotations)
LNG1_standard$annotation <- factor(LNG1_standard$annotation, levels = selected_annotations)

# Merge data
LNG_tot_all <- MergeSTData(LNG1_RRST, LNG1_standard)

# Keep 100 spots per protocol and annotation
keep_spots <- LNG_tot_all[[]] |> 
  rownames_to_column(var = "barcode") |> 
  select(annotation, protocol, barcode) |> 
  group_by(annotation, protocol) |> 
  slice_sample(n = 25) |> 
  pull(barcode)
LNG_tot <- SubsetSTData(LNG_tot_all, spots = keep_spots) |> 
  NormalizeData() |> 
  ScaleData()
LNG_tot$annotation <- factor(LNG_tot$annotation, levels = selected_annotations)
LNG_tot <- SetIdent(LNG_tot, value = "annotation")

# plot DEG heatmap
p <- DoHeatmap(LNG_tot, features = shared_markers$gene, raster = TRUE, disp.min = -2.5, disp.max = 2.5) +
  theme(axis.text.y = element_text(size = 5, colour = "black")) +
  scale_fill_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") |> rev())
pdf(file = "../Suppl_figures/Suppl_Figure_lung_markers/DE_heatmap_protocol_comparison.pdf", width = 9, height = 14)
print(p)
dev.off()

# Detection rates
p_det_rate <- ggplot(shared_markers, aes(pct_RRST, pct_standard, 
                                         color = ifelse(pct_RRST > pct_standard, "RRST", "standard"))) +
  geom_abline(slope = 1, intercept = 0, linetype = "longdash") +
  geom_point() +
  theme_classic() +
  scale_x_continuous(limits = c(0, 1)) +
  scale_y_continuous(limits = c(0, 1)) +
  labs(color = "detection rate\nhigher in:") +
  facet_grid(annotation ~ .) +
  theme(panel.background = element_rect(colour = "black"))

pdf(file = "../Suppl_figures/Suppl_Figure_lung_markers/det_rate_marker_genes.pdf", height = 8, width = 4)
print(p_det_rate)
dev.off()

# Export cluster QC plot
pdf(file = "../Suppl_figures/Suppl_Figure_lung_markers/unique_genes_clusters.pdf", height = 2.5, width = 7)
LNG_tot_all[[]] |> 
  ggplot(aes(annotation, nFeature_RNA, fill = protocol)) +
  geom_violin(scale = "width", position = "dodge") +
    geom_point(position=position_jitterdodge(dodge.width=0.9), size = 0.5) +
    theme_classic() +
    theme(axis.text.x = element_blank()) +
    labs(y = "Unique genes", x = "")
dev.off()

```


# date
***

```{r}
date()
```

# Session
***

```{r}
devtools::session_info()
```