---
title: "Figure 2, Suppl. Figure 2 and 3"
author: "Ludvig Larsson"
date: '2022-05-23'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load libraries

```{r eval=FALSE}

library(STutility)
library(ggplot2)
library(ggpubr)

```

Assemble spaceranger output files and merge curated meta data

```{r eval=FALSE}

# Mouse brain
samples <- Sys.glob(paths = "../../spaceranger_output/lung/*/filtered_feature_bc_matrix.h5")
imgs <- Sys.glob(paths = "../../spaceranger_output/lung/*/spatial/tissue_hires_image.png")
spotfiles <- Sys.glob(paths = "../../spaceranger_output/lung/*/spatial/tissue_positions_list.csv")
json <- Sys.glob(paths = "../../spaceranger_output/lung/*/spatial/scalefactors_json.json")

infoTable <- data.frame(samples, imgs, spotfiles, json)
infoTable <- cbind(infoTable, arrayID = do.call(rbind, strsplit(infoTable$samples, "/"))[, 5])

curated_metadata <- openxlsx::read.xlsx("../../sheets/curated_RNA_rescue_sample_metadata.xlsx", sheet = 3)
curated_metadata <- setNames(curated_metadata, nm = c("storage_time", "seq_date", "include", "RNA_rescue",
                                                       "project", "experimenter", "processer", "comments",
                                                       "ID", "paper_id", "protocol", "source", "arrayID", "spots_under_tissue",
                                                       "genes_detected", "fraction_spots_under_tissue",
                                                       "median_genes_per_spot", "median_UMIs_per_spot", 
                                                       "saturation", "reads_mapped_to_probe_set",
                                                       "reads_mapped_confidently_to_probe_set",
                                                       "reads_mapped_confidently_to_filtered_probe_set",
                                                       "reads_mapped_to_genome",
                                                       "reads_mapped_confidently_to_genome",
                                                       "number_of_panel_genes"))

infoTable <- merge(infoTable, curated_metadata, by = "arrayID")

```

Load data into a Seurat object

* 1 : LNG1, sRR-seq 
* 2-3 : LNG1, standard
* 4 : LNG2, sRR-seq 
* 5-6 : LNG2, standard

```{r eval=FALSE}

LNG <- InputFromTable(infoTable[c(1, 3:4, 2, 5:6), ])

```

```{r eval=FALSE}

ST.FeaturePlot(LNG, features = "nFeature_RNA", ncol = 3)

```

Load images

```{r eval=FALSE}

LNG <- LoadImages(LNG, time.resolve = FALSE, xdim = 1e3)

```

Add empty masks for downstream processing of HE images

```{r eval=FALSE}

# Prepare data for rigid transformation
st.object <- GetStaffli(LNG)
msks <- lapply(st.object@rasterlists$raw, function(im) {
  as.raster(matrix(data = 1, ncol = ncol(im), nrow = nrow(im)))
})
st.object@rasterlists$masked <- st.object@rasterlists$raw
st.object@rasterlists$masked.masks <- msks
LNG@tools$Staffli <- st.object

```

Apply transformations

```{r eval=FALSE}

# Warp transform
LNG <- WarpImages(LNG, verbose = TRUE, transforms = list( "2" = list(angle = -90), "3" = list(angle = -90), "5" = list(angle = 180)))

```

Add a new metadata column with a combined protocol and lung ID label

```{r eval=FALSE}

LNG$protocol_lng <- gsub(pattern = " ", replacement = "_", paste0(LNG$protocol, "_ID: ", LNG$paper_id))
ST.FeaturePlot(LNG, features = "nFeature_RNA", ncol = 3, label.by = "protocol_lng")

```

## QC
***

The standard Visium data sets have reduced numbers of unique genes compared to the data generated by RR-seq.

```{r eval=FALSE}

VlnPlot(LNG, features = "nFeature_RNA", group.by = "arrayID", split.by = "protocol")

```

At a first glance, we can see that the distribution of unique genes is quite different across the two protocols. For RR-seq, we have much more even coverage across the tissue, whereas in the standard Visium data, we only get high counts in the epithelium. 

A reasonable threshold for filtering would be somewhere between 300-1000 unique genes, depending on the tissue type. If parts of the tissue is cell sparse, which is the case for the alveoli, it might be more reasonable to set a lower threshold. 

If we use a threshold of 300 unique genes, we can see that ~1.3%-2.3% of spots are discarded from the RR-seq data, whereas ~21%-80% of spots are discarded from the standard Visium data.

```{r eval=FALSE}

LNG[[]] %>% 
  dplyr::mutate(filter = ifelse(nFeature_RNA < 300, "discard", "keep")) %>%
  dplyr::group_by(protocol, paper_id, filter) %>%
  dplyr::summarize(Freq = n()) %>%
  dplyr::group_by(protocol, paper_id) %>%
  dplyr::mutate(percentage_discarded = round(Freq/sum(Freq)*100, digits = 2)) %>%
  dplyr::filter(filter == "discard") %>%
  dplyr::select(-Freq, -filter) %>%
  dplyr::arrange(paper_id)

p <- LNG[[]] %>% ggplot(aes(nFeature_RNA, fill = ifelse(nFeature_RNA < 300, "discard", "keep"))) +
  geom_histogram(binwidth = 100) +
  geom_vline(xintercept = 300, linetype = "longdash") +
  facet_grid(protocol~ID, scales = "free", 
             labeller = labeller(protocol = c("RNA rescue" = "sRR-seq", "standard" = "standard"), 
                                 ID = c("P583-B-LNG3" = "ID: 1", "P640-C-LNG5" = "ID: 2"))) +
  labs(fill = "", x = "Unique genes", y = "Count") +
  scale_x_continuous(breaks = c(0, 300, 2000, 4000, 6000), minor_breaks = NULL) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1))

pdf(file = "../Suppl_Figure_2/Suppl_Figure_2.pdf", width = 12, height = 8)
print(p)
dev.off()

png(filename = "../Suppl_Figure_2/Suppl_Figure_2.png", width = 2400, height = 1600, res = 300)
print(p)
dev.off()

```

### Filter data
***

```{r eval=FALSE}

LNG$filter <- ifelse(LNG$nFeature_RNA < 300, "discard", "keep")

LNG.subset <- SubsetSTData(LNG, expression = nFeature_RNA > 300)

```

Now we can see that quite a few spots have been discarded, in particular for the standard Visium data for ID 2.

```{r eval=FALSE}

ST.FeaturePlot(LNG.subset, features = "nFeature_RNA", ncol = 3, label.by = "protocol_lng")

```


Split data by protocol and sample

```{r eval=FALSE}

LNGList <- lapply(unique(LNG.subset$protocol_lng), function(i) {
  SubsetSTData(LNG.subset, expression = protocol_lng == i)
})

```

## Analysis
***

1. log-normalize data
2. scale data (z-transformation)
3. find top variable features
4. run dimensionality reduction using Principal Component Analysis (PCA)
5. Create an SNN graph using the first 20 PCs as input
6. Find clusters using Louvain algorithm
7. Create a UMAP using the first 20 PCs as input

```{r eval=FALSE}

LNGList <- lapply(LNGList, function(LNG.subset) {
  LNG.subset <- LNG.subset %>%
    NormalizeData() %>%
    ScaleData() %>%
    FindVariableFeatures() %>%
    RunPCA() %>%
    FindNeighbors(reduction = "pca", dims = 1:20) %>%
    FindClusters(resolution = 0.8) %>%
    RunUMAP(reduction = "pca", dims = 1:20, min.dist = 0.3, n.epochs = 1e3)
})
LNGList <- setNames(LNGList, nm = unique(LNG$protocol_lng))

```

Plot clusters in UMAP embedding

```{r eval=FALSE}

DimPlot(LNGList[[1]]) | DimPlot(LNGList[[3]])
DimPlot(LNGList[[2]]) | DimPlot(LNGList[[4]])

```

## DEA
***

Run DEA with an avg_log2FC threshold of 0.5 and adjusted p-value < 0.01.

```{r eval=FALSE}

# Keep lower threshold avg_log2FC for checking DE genes in all clusters
de.markers.list.025 <- lapply(LNGList, function(se) {
  se <- SetIdent(se, value = "seurat_clusters")
  FindAllMarkers(se, only.pos = TRUE, logfc.threshold = 0.25)
})
de.markers.list.025 <- lapply(de.markers.list.025, function(de.markers) {
  de.markers %>% dplyr::filter(p_val_adj < 0.01)
})

# avg_log2FC > 0.5 
# These are the DE genes used for figure plots
de.markers.list <- lapply(de.markers.list.025, function(de.markers) {
  de.markers %>% dplyr::filter(avg_log2FC > 0.5)
})


```

### Add cluster annotations
***

```{r}

# sRR-seq cluster annotations
ann1 <- c("0" = "AT1-enriched", "1" = "Macrophage-enriched",
         "2" = "AT2-enriched", "3" = "Connective tissue 1", 
         "4" = "Connective tissue 2", "5" = "Connective tissue 3",
         "6" = "Airway epithelium", "7" = "Megakaryocyte-enriched",
         "8" = "Smooth muscle", "9" = "Immune-enriched", 
         "10" = "Glands")
ann3 <- c("0" = "AT2-enriched", "1" = "Low-quality cluster 1",
          "2" = "Connective tissue 1", "3" = "Connective tissue 2",
          "4" = "Low-quality cluster 2", "5" = "Connective tissue 3",
          "6" = "Glands", "7" = "Megakaryocyte-enriched", 
          "8" = "Airway epithelium")

```


### Create plots for figure 2
***

* plots_vln : violin plots showing the number of unique genes and UMIs 
* plots_filter : spatial plots showing what spots are discarded in the filtering step (unique genes < 300)
* plots_umap : UMAP plots colored by cluster
* plots_qc : spatial plots showing the number of unique genes
* plots_HE : HE images
* plots_spatial : spatial plots showing clusters, split by panel
* plots_dotplot : dot plot showing DE genes per cluster

```{r eval=FALSE}

# Violin plots showing the number of unique genes and UMIs
plots_vln <- lapply(1:4, function(i) {
  p <- LNGList[[i]][[]] %>%
    dplyr::mutate(`Unique genes` = nFeature_RNA, `UMI counts` = nCount_RNA) %>%
    reshape2::melt(measure.vars = c("Unique genes", "UMI counts")) %>%
    ggplot(aes(paper_id, value, fill = variable)) +
    geom_violin(scale = "width", alpha = 0.7) +
    geom_boxplot(width = 0.3) +
    theme_classic() +
    facet_wrap(~variable, scales = "free") +
    theme(axis.title = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
          legend.position = "none", axis.text.y = element_text(size = 18, colour = "black"), strip.text = element_text(size = 18))
  return(p)
})

# Spatial plots showing the spots discarded
plots_filter <- lapply(c("RNA_rescue_ID:_1.0", "standard_ID:_1.0", "RNA_rescue_ID:_2.0", "standard_ID:_2.0"), function(i) {
  p <- ST.FeaturePlot(SubsetSTData(LNG, expression = protocol_lng %in% i), features = "filter", show.sb = FALSE, pt.size = 1.3, 
                 cols = c("keep" = "lightgray", "discard" = "red"), indices = ifelse(i %in% c("standard_ID:_1.0", "standard_ID:_2.0"), 2, 1)) &
    theme(strip.text = element_blank(), plot.title = element_blank(), legend.title = element_blank(), legend.text = element_text(size = 16)) &
  guides(fill = guide_legend(override.aes = list(size = 7)))
  p <- ggrastr::rasterize(p, layers = "Point", dpi = 300)
  return(p)
})

# UMAP plots showing clusters
plots_umap <- lapply(1:4, function(i) {
  gg <- cbind(LNGList[[i]][[]], LNGList[[i]]@reductions$umap@cell.embeddings)

  p <- ggplot() +
    geom_point(data = gg, aes(UMAP_1, UMAP_2, color = seurat_clusters)) +
    ggrepel::geom_label_repel(size = 9, label.size = 0.2, data = gg %>% 
                 dplyr::group_by(seurat_clusters) %>% 
                 summarize(med1 = median(UMAP_1), med2 = median(UMAP_2)), aes(med1, med2 + 0.6, fill = seurat_clusters, label = seurat_clusters)) +
    theme_classic() +
    theme(plot.title = element_blank(), 
          legend.position = "none", axis.text = element_text(size = 14, colour = "black"), axis.title = element_text(size = 18))
  p <- ggrastr::rasterize(p, layers = "Point", dpi = 300)
  return(p)
})

# Spatial plots showing the number of unique genes
plots_qc <- lapply(1:4, function(i) {
  p <- ST.FeaturePlot(LNGList[[i]], features = "nFeature_RNA", show.sb = FALSE, pt.size = 1.3, 
                      indices = ifelse(i %in% c(2, 4), 2, 1), cols = viridis::magma(n = 11, direction = -1)) &
    theme(plot.subtitle = element_blank(), 
          plot.title = element_blank(), 
          strip.text = element_blank(), 
          legend.text = element_text(size = 14), 
          legend.title = element_text(size = 18)) &
    labs(fill = "Unique\ngenes") 
  p <- ggrastr::rasterize(p, layers = "Point", dpi = 300)
  return(p)
})

# HE images
plots_HE <- lapply(1:4, function(i) {
  im <- GetStaffli(LNGList[[i]])@rasterlists$processed[[ifelse(i %in% c(2, 4), 2, 1)]]
  im[, 1:100] <- "#ffffffff"; im[, (ncol(im) - 120):ncol(im)] <- "#ffffffff"
  im <- im[101:(nrow(im) - 100), ]
  g <- grid::rasterGrob(im, width = unit(1, "npc"), height = unit(1, "npc"), interpolate = TRUE)
  ggplot() +
    annotation_custom(g, -Inf, Inf, -Inf, Inf) +
    theme_void() +
    theme(plot.margin = margin(t = 0, r = 40, b = 0, l = 40))
})

# Spatial plots showing clusters split by clusters
plots_spatial <- lapply(1:4, function(i) {
  p <- ST.FeaturePlot(LNGList[[i]], features = "seurat_clusters", split.labels = TRUE, ncol = ifelse(i %in% c(1, 2), 4, 3), 
                 indices = ifelse(i %in% c(2, 4), 2, 1), show.sb = FALSE, pt.size = 0.5) +
    theme(plot.title = element_blank(), legend.position = "none", strip.text = element_blank()) &
    scale_x_continuous(limits = c(150, 1850)) &
    scale_y_continuous(limits = c(150, 1850))
  p <- ggrastr::rasterize(p, layers = "Point", dpi = 300)
  return(p)
})

# Dot plot showing DE genes
plots_dotplot <- lapply(1:4, function(i) {
  LNG <- LNGList[[i]]
  if (i == 1) {
    LNG$seurat_clusters <- factor(ann1[LNG$seurat_clusters], levels = ann1)
  }
  if (i == 2) {
    LNG$seurat_clusters <- factor(ann3[LNG$seurat_clusters], levels = ann3)
  }
  LNG <- Seurat::SetIdent(LNG, value = "seurat_clusters")
  p <- DotPlot(LNG, features = unique(de.markers.list[[i]] %>% 
                                          dplyr::mutate(gene_count = table(gene)[gene]) %>%
                                          dplyr::filter(gene_count == 1) %>%
                                          dplyr::group_by(cluster) %>%
                                          dplyr::slice_head(n = 4) %>% 
                                          dplyr::pull(gene) %>% rev())) +
    scale_color_gradientn(colours = scico::scico(n = 11, palette = "bam", direction = -1)) +
    scale_y_discrete(position = "left") +
    theme(axis.title = element_blank(), axis.text = element_text(size = 14), legend.text = element_text(size = 14), 
          legend.title = element_text(size = 14), axis.text.x = element_text(angle = 60, hjust = 1, size = 16)) +
    coord_flip()
  p$guides$colour$title <- "Avg. Expr."
  p$guides$size$title <- "Pct. Expr."
  return(p)
})

```

### Patchwork for Figure 2 
***

This patchwork includes only LNG1 and each plot type is places side by side for a simple comparison between the two protocols.

```{r}

# Design for patchwork
design <- c(
  patchwork::area(1, 1, 1, 1),
  patchwork::area(1, 2, 1, 2),
  patchwork::area(2, 1, 2, 1),
  patchwork::area(2, 2, 2, 2),
  patchwork::area(3, 1, 3, 1),
  patchwork::area(4, 1, 4, 1),
  patchwork::area(3, 2, 3, 2),
  patchwork::area(4, 2, 4, 2),
  
  patchwork::area(1, 3, 1, 3),
  patchwork::area(1, 4, 1, 4),
  patchwork::area(2, 3, 2, 3),
  patchwork::area(2, 4, 2, 4),
  patchwork::area(3, 3, 5, 3),
  patchwork::area(3, 4, 5, 4)
)

# Create patchwork
p <- plots_HE[[1]] +
  plots_HE[[2]] +
  plots_qc[[1]] +
  plots_qc[[2]] +
  plots_umap[[1]] + 
  plots_spatial[[1]] +
  plots_umap[[2]] +
  plots_spatial[[2]] +
  plots_vln[[1]] +
  plots_vln[[2]] +
  plots_filter[[1]] +
  plots_filter[[2]] +
  plots_dotplot[[1]] +
  plots_dotplot[[2]] +
  patchwork::plot_layout(design = design, heights = c(1.5, 2, 2, 2, 0.5))

pdf(file = "figure_2_patchwork_side_by_side.pdf", width = 26, height = 20)
print(p)
dev.off()

```

### Patchwork for SUpplementary Figure 
***

This patchwork includes only LNG2 and each plot type is places side by side for a simple comparison between the two protocols.

```{r}

# Design for patchwork
design <- c(
  patchwork::area(1, 1, 1, 1),
  patchwork::area(1, 2, 1, 2),
  patchwork::area(2, 1, 2, 1),
  patchwork::area(2, 2, 2, 2),
  patchwork::area(3, 1, 3, 1),
  patchwork::area(4, 1, 4, 1),
  patchwork::area(3, 2, 3, 2),
  patchwork::area(4, 2, 4, 2),
  
  patchwork::area(1, 3, 1, 3),
  patchwork::area(1, 4, 1, 4),
  patchwork::area(2, 3, 2, 3),
  patchwork::area(2, 4, 2, 4),
  patchwork::area(3, 3, 5, 3),
  patchwork::area(3, 4, 5, 4)
)

# Create patchwork
p <- plots_HE[[3]] +
  plots_HE[[4]] +
  plots_qc[[3]] +
  plots_qc[[4]] +
  plots_umap[[3]] + 
  plots_spatial[[3]] +
  plots_umap[[4]] +
  plots_spatial[[4]] +
  plots_vln[[3]] +
  plots_vln[[4]] +
  plots_filter[[3]] +
  plots_filter[[4]] +
  plots_dotplot[[3]] +
  plots_dotplot[[4]] +
  patchwork::plot_layout(design = design, heights = c(1.5, 2, 2, 2, 0.5))

pdf(file = "suppl_figure_lung_patchwork_side_by_side.pdf", width = 26, height = 20)
print(p)
dev.off()

```





### Patchwork for Figure 2 with both LNG1 and LNG2

```{r eval=FALSE}

# Design for patchwork
design <- c(
  patchwork::area(1, 1, 1, 1),
  patchwork::area(1, 2, 1, 2),
  patchwork::area(2, 1, 2, 1),
  patchwork::area(2, 2, 2, 2),
  patchwork::area(3, 1, 3, 1),
  patchwork::area(4, 1, 4, 1),
  patchwork::area(3, 2, 4, 2),
  
  patchwork::area(1, 3, 1, 3),
  patchwork::area(1, 4, 1, 4),
  patchwork::area(2, 3, 2, 3),
  patchwork::area(2, 4, 2, 4),
  patchwork::area(3, 3, 3, 3),
  patchwork::area(4, 3, 4, 3),
  patchwork::area(3, 4, 4, 4)
)

# Create patchwork
p <- plots_vln[[1]] +
  plots_HE[[1]] +
  plots_qc[[1]] +
  plots_filter[[1]] +
  plots_spatial[[1]] +
  plots_umap[[1]] + 
  plots_dotplot[[1]] +
  plots_vln[[2]] +
  plots_HE[[2]] +
  plots_qc[[2]] +
  plots_filter[[2]] +
  plots_spatial[[2]] +
  plots_umap[[2]] +
  plots_dotplot[[2]] +
  patchwork::plot_layout(design = design, heights = c(1.5, 2, 2, 2))

pdf(file = "figure_2_patchwork_v2.pdf", width = 26, height = 17)
print(p)
dev.off()

```

### Patchwork for Figure 2 with LNG1 only

```{r eval=FALSE}

# Create patchwork
p <- plots_vln[[3]] +
  plots_HE[[3]] +
  plots_qc[[3]] +
  plots_filter[[3]] +
  plots_spatial[[3]] +
  plots_umap[[3]] + 
  plots_dotplot[[3]] +
  plots_vln[[4]] +
  plots_HE[[4]] +
  plots_qc[[4]] +
  plots_filter[[4]] +
  plots_spatial[[4]] +
  plots_umap[[4]] +
  plots_dotplot[[4]] +
  patchwork::plot_layout(design = design, heights = c(1.5, 2, 2, 2))

pdf(file = "figure_2_patchwork_ID2_v2.pdf", width = 26, height = 17)
print(p)
dev.off()

```

### Alternative Figure 2 layout, LNG1 and LNG2 separate

```{r eval=FALSE}

# Dot plot showing DE genes
plots_dotplotv2 <- lapply(1:4, function(i) {
  p <- DotPlot(LNGList[[i]], features = unique(de.markers.list[[i]] %>% 
                                          dplyr::mutate(gene_count = table(gene)[gene]) %>%
                                          dplyr::filter(gene_count == 1) %>%
                                          dplyr::group_by(cluster) %>%
                                          dplyr::slice_head(n = 6) %>% 
                                          dplyr::pull(gene))) +
    scale_color_gradientn(colours = scico::scico(n = 11, palette = "bam", direction = -1)) +
    scale_y_discrete(position = "right") +
    theme(axis.title = element_blank(), 
          axis.text = element_text(size = 12), 
          axis.text.x = element_text(angle = 90, hjust = 1),
          legend.text = element_text(size = 12), 
          legend.title = element_text(size = 14))
  p$guides$colour$title <- "Avg. Expr."
  p$guides$size$title <- "Pct. Expr."
  return(p)
})

# Design for patchwork
design <- c(
  patchwork::area(1, 1, 1, 1),
  patchwork::area(1, 2, 1, 2),
  patchwork::area(2, 1, 2, 1),
  patchwork::area(2, 2, 2, 2),
  patchwork::area(1, 3, 2, 4),
  patchwork::area(3, 1, 3, 1),
  patchwork::area(3, 2, 3, 4)
)

# Create patchwork 1
p <- plots_vln[[1]] +
  plots_HE[[1]] +
  plots_qc[[1]] +
  plots_filter[[1]] +
  plots_spatial[[1]] +
  plots_umap[[1]] + 
  plots_dotplotv2[[1]] +
  patchwork::plot_layout(design = design, heights = c(1.3, 2, 2))

pdf(file = "figure_2_patchwork_v2.pdf", width = 18, height = 12)
print(p)
dev.off()

# Create patchwork 2
p <- plots_vln[[2]] +
  plots_HE[[2]] +
  plots_qc[[2]] +
  plots_filter[[2]] +
  plots_spatial[[2]] +
  plots_umap[[2]] + 
  plots_dotplotv2[[2]] +
  patchwork::plot_layout(design = design, heights = c(1.3, 2, 2))

pdf(file = "figure_2_patchwork_ID2_v2.pdf", width = 18, height = 12)
print(p)
dev.off()

```

### Supplementary figure 3

```{r eval=FALSE}

# sRR-seq cluster annotations
ann <- c("0" = "AT1-enriched", "1" = "Macrophage-enriched",
         "2" = "AT2-enriched", "3" = "Connective tissue 1", 
         "4" = "Connective tissue 2", "5" = "Connective tissue 3",
         "6" = "Airway epithelium", "7" = "Megakaryocyte-enriched",
         "8" = "Smooth muscle", "9" = "Immune-enriched", 
         "10" = "Glands")
ann2 <- c("0" = "AT2-enriched", "1" = "Low-quality cluster 1",
          "2" = "Connective tissue 1", "3" = "Connective tissue 2",
          "4" = "Low-quality cluster 2", "5" = "Connective tissue 3",
          "6" = "Glands", "7" = "Megakaryocyte-enriched", 
          "8" = "Airway epithelium")
lvls <- c(ann, c("Low-quality cluster 1", "Low-quality cluster 2"))

# Dot plot showing DE genes
LNGList[[1]]$ann <- factor(ann[LNGList[[1]]$seurat_clusters], levels = ann)
LNGList[[1]] <- SetIdent(LNGList[[1]], value = "ann")
dotplot_LNG1 <- DotPlot(LNGList[[1]], features = unique(de.markers.list[[i]] %>% 
                                          dplyr::mutate(gene_count = table(gene)[gene]) %>%
                                          dplyr::filter(gene_count == 1) %>%
                                          dplyr::group_by(cluster) %>%
                                          dplyr::slice_head(n = 6) %>% 
                                          dplyr::pull(gene))) +
  scale_color_gradientn(colours = scico::scico(n = 11, palette = "bam", direction = -1)) +
  scale_y_discrete(position = "right") +
  theme(axis.title = element_blank(), 
        axis.text = element_text(size = 12), 
        axis.text.x = element_text(angle = 90, hjust = 1),
        legend.text = element_text(size = 12), 
        legend.title = element_text(size = 14))
dotplot_LNG1$guides$colour$title <- "Avg. Expr."
dotplot_LNG1$guides$size$title <- "Pct. Expr."


selected.markers.airway.epithelium <- c("CFAP99", "DNAI2", "CFAP65", "CCDC33", "CCDC65")
selected.markers.glands <- c("LPO", "GCNT3", "C2CD4A", "TCN1", "PIP")
selected.markers.platelet <- c("GFI1B", "GP9", "MFSD2B", "GATA1", "GP1BA")
selected.markers.smoothmuscle <- c("LGR6", "PLN", "HSPB7", "KCNA5", "CSDC2", "SCUBE3")

all.markers <- list("Airway epithelium" = selected.markers.airway.epithelium, 
                    "Glands" = selected.markers.glands, 
                    "Megakaryocyte-enriched" = selected.markers.platelet, 
                    "Smooth muscle" = selected.markers.smoothmuscle)

LNGList[[1]] <- SetIdent(LNGList[[1]], value = "seurat_clusters")
ggm <- do.call(rbind, lapply(names(all.markers), function(n) {
  print(n)
  mrks <- all.markers[[n]][1:5]
  data1 <- cbind(FetchData(LNGList[[1]], slot = "data", vars = mrks), type = "sRR-seq")
  data2 <- cbind(FetchData(LNGList[[2]], slot = "data", vars = mrks), type = "standard")
  data <- rbind(data1, data2)
  gg <- cbind(rbind(LNGList[[1]][[]] %>% dplyr::select(-ann), LNGList[[2]][[]]), data)
  ggm <- reshape2::melt(gg, measure.vars = mrks)
  ggm$region <- n
  return(ggm)
}))

plot_selected_markers <- ggm %>% 
  dplyr::mutate(annotations = ifelse(type == "sRR-seq", ann[seurat_clusters], ann2[seurat_clusters])) %>%
  dplyr::mutate(annotations = factor(annotations, levels = lvls), variable = 
                  factor(variable, levels = unlist(all.markers[c(1, 3, 4, 2)]))) %>%
  ggplot(aes(annotations, value, fill = seurat_clusters)) +
    geom_violin(scale = "width") +
    geom_jitter(size = 0.1) +
    theme_classic() +
    facet_grid(variable ~ type, space = "free", scales = "free_x") +
    theme(legend.position = "none", 
          axis.text.x = element_text(angle = 90, hjust = 1, size = 12),
          axis.text.y = element_text(size = 12),
          strip.background = element_rect(colour = NA), 
          strip.text.x = element_blank(),
          strip.text.y = element_text(angle = 0, hjust = 0), 
          axis.title.y = element_text(size = 14)) +
    labs(y = "Log-normalized expression", x = "")

plot_selected_markers <- ggrastr::rasterize(plot_selected_markers , layers = "Point", dpi = 300)

# Create patchwork
design <- c(patchwork::area(1, 1, 1, 2),
            patchwork::area(2, 1, 2, 2))
p <- dotplot_LNG1 + 
  plot_selected_markers + 
  patchwork::plot_layout(nrow = 2, heights = c(1, 4.5), design = design)

pdf(file = "../Suppl_Figure_3/Suppl_figure_3_patchwork_markers.pdf", width = 15, height = 18)
print(p)
dev.off()

```


# date
***

```{r eval=FALSE}
date()
```

# Session
***

```{r eval=FALSE}
devtools::session_info()
```