---
title: "Figure 1"
author: "Ludvig Larsson"
date: '2022-05-23'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, root.dir = "/home/rstudio")
```

Load required libraries

```{r}

library(STutility)
library(ggplot2)
library(ggpubr)
library(ggrastr)
library(dplyr)
library(magrittr)
library(ggbreak)

```


Download files from Mendeley.

```{r eval=FALSE}

# Create data directories
suppressWarnings({
  dir.create("../data/")
  dir.create("../data/genes/")
  dir.create("../data/sheets/")
  dir.create("../data/spaceranger_output/")
})

source("../scripts/mendeley.R")

# Download gene annotation files
for (i in seq_along(genes.files)) {
  download.file(url = genes.files[i], destfile = paste0("../data/genes/", names(genes.files)[i]))
}

# Download gene annotation files
for (i in seq_along(sheets.files)) {
  download.file(url = sheets.files[i], destfile = paste0("../data/sheets/", names(sheets.files)[i]))
}

# Download spaceranger output files
for (i in 5:6) {
  download.file(url = spaceranger.files[i], destfile = paste0("../data/spaceranger_output/", names(spaceranger.files)[i]))
  untar(tarfile = paste0("../data/spaceranger_output/", names(spaceranger.files)[i]), exdir = "../data/spaceranger_output/")
  file.remove(paste0("../data/spaceranger_output/", names(spaceranger.files)[i]))
}

```


Assemble spaceranger output files


```{r}

# Mouse brain
samples1 <- Sys.glob(paths = "../data/spaceranger_output/mousebrain/*/filtered_feature_bc_matrix.h5")
imgs1 <- Sys.glob(paths = "../data/spaceranger_output/mousebrain/*/spatial/tissue_hires_image.png")
spotfiles1 <- Sys.glob(paths = "../data/spaceranger_output/mousebrain/*/spatial/tissue_positions_list.csv")
json1 <- Sys.glob(paths = "../data/spaceranger_output/mousebrain/*/spatial/scalefactors_json.json")

# Prostate cancer
samples2 <- Sys.glob(paths = "../data/spaceranger_output/prostatecancer/*/filtered_feature_bc_matrix.h5")
imgs2 <- Sys.glob(paths = "../data/spaceranger_output/prostatecancer/*/spatial/tissue_hires_image.png")
spotfiles2 <- Sys.glob(paths = "../data/spaceranger_output/prostatecancer/*/spatial/tissue_positions_list.csv")
json2 <- Sys.glob(paths = "../data/spaceranger_output/prostatecancer/*/spatial/scalefactors_json.json")

infoTable <- data.frame(samples = c(samples1, samples2), imgs = c(imgs1, imgs2), spotfiles = c(spotfiles1, spotfiles2), json = c(json1, json2))
infoTable <- cbind(infoTable, arrayID = do.call(rbind, strsplit(infoTable$samples, "/"))[, 5])

curated_metadata <- openxlsx::read.xlsx("../data/sheets/RRST_sample_metadata.xlsx")
curated_metadata <- setNames(curated_metadata, nm = c("storage_time", "seq_date", "comments",
                                                       "ID", "paper_id", "RIN", "DV200", "protocol", "source", 
                                                       "arrayID", "spots_under_tissue",
                                                       "genes_detected", "fraction_spots_under_tissue",
                                                       "median_genes_per_spot", "median_UMIs_per_spot", 
                                                       "saturation", "reads_mapped_to_probe_set",
                                                       "reads_mapped_confidently_to_probe_set",
                                                       "reads_mapped_confidently_to_filtered_probe_set",
                                                       "reads_mapped_to_genome",
                                                       "reads_mapped_confidently_to_genome",
                                                       "number_of_panel_genes"))

infoTable <- merge(infoTable, curated_metadata, by = "arrayID")

```

Load mouse brain Visium data (4xRRST + 4xstandard)

```{r}

# Subset infoTable to include mosuebrain data
MB_infoTable <- subset(infoTable, source == "mouse brain")
MB <- InputFromTable(infotable = MB_infoTable)

```

Load H&E images into the `MB` object

```{r fig.width=8, fig.height=4}

MB <- LoadImages(MB, time.resolve = FALSE)
ImagePlot(MB, ncols = 4)

```

Here we apply some rotations to make a rough alignment of the H&E images.

```{r}

# Warp transform
MB <- WarpImages(MB, verbose = TRUE, transforms = list("1" = list(angle = 90), "2" = list(angle = -90), 
                                                       "3" = list(angle = 90), "4" = list(angle = 90),
                                                       "5" = list(angle = -90), "6" = list(angle = -90), 
                                                       "7" = list(angle = 90), "8" = list(angle = -90)))
```

# Figure 1
***

## b - Spatial map of unique genes per spot [mousebrain]
***

```{r fig.width=8, fig.height=5}

p <- ST.FeaturePlot(MB, features = "nFeature_RNA", ncol = 2, indices = c(1, 5), show.sb = FALSE,
                    label.by = "protocol",  pt.size = 1.2) &
  theme(plot.title = element_blank(), legend.position = "top", 
        legend.text = element_text(angle = 60, hjust = 1, size = 14, colour = "black"),
        legend.title = element_text(size = 18, face = "bold", colour = "black", vjust = 1),
        strip.text = element_text(size = 18, face = "bold", colour = "black")) &
  labs(fill = "unique genes")
p

```

```{r eval=FALSE}
jpeg(filename = "plots/mousebrain_unique_genes_spatial.jpg", width = 2200, height = 1400, res = 300)
print(p)
dev.off()

```

## c - Violin plots of unique genes per spot [mousebrain]
***

```{r}

p <- ggplot(MB[[]], aes(arrayID, nFeature_RNA, fill = protocol)) +
  geom_violin(scale = "width") +
  geom_boxplot(width = 0.3) +
  geom_hline(data = MB[[]] %>% dplyr::group_by(protocol) %>% summarize(Median = median(nFeature_RNA)), 
             aes("", yintercept = Median, group = protocol), linetype = "longdash") +
  geom_label(data = MB[[]] %>% dplyr::group_by(protocol) %>% summarize(Median = median(nFeature_RNA)), size = 5,
             aes("", y = Median, group = protocol, label = Median)) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1, size = 14, colour = "black"), 
        axis.text.y = element_text(size = 14, colour = "black"), 
        strip.text = element_text(size = 18, face = "bold", colour = "black")) +
  labs(y = "", x = "") +
  facet_grid(~protocol, scales = "free") +
  guides(fill = "none") +
  scale_x_discrete(labels = c("", "Rep1", "Rep2", "Rep3", "Rep4"))

p

```

```{r eval=FALSE}

# Export plot
jpeg(filename = "plots/mousebrain_unique_genes_violin.jpg", width = 2000, height = 900, res = 300)
print(p)
dev.off()

```

Calculate gene attributes for mouse brain data

```{r}

gene_attr <- do.call(cbind, lapply(c("RRST", "standard"), function(pro) {
  umis <- GetAssayData(MB, slot = "counts", assay = "RNA")[, MB$protocol %in% pro]
  gene_attr_protocol <- data.frame(rowSums(umis), rowMeans(umis > 0), row.names = rownames(umis))
  gene_attr_protocol <- setNames(gene_attr_protocol, nm = c(paste0(gsub(pattern = " ", replacement = "_", x = pro), "_count"), 
                                                            paste0(gsub(pattern = " ", replacement = "_", x = pro), "_det_rate")))
  return(gene_attr_protocol)
}))

gene_attr <- gene_attr %>%
  mutate("RRST_log_count" = log1p(`RRST_count`), "standard_log_count" = log1p(`standard_count`))

```

Subset genes to include those available in the Visium FFPE probe set.

```{r}

mouse_probes <- read.csv("../data/sheets/Visium_Mouse_Transcriptome_Probe_Set_v1.0_mm10-2020-A.csv", skip = 5, header = TRUE)
mouse_probes$gene_name <- do.call(rbind, strsplit(mouse_probes$probe_id, "\\|"))[, 2]

```

## d - Gene-gene scatter plot of detection rate and UMI counts [mouse brain]
***

```{r fig.width=8, fig.height=4}

p1 <- ggplot(gene_attr %>% mutate(gene = rownames(.)) %>% 
               subset(gene %in% mouse_probes$gene_name), 
             aes_string("standard_log_count", "RRST_log_count")) +
  geom_point(size = 0.3) +
  ggpubr::stat_cor(digits = 2) +
  scale_x_continuous(limits = c(0, max(max(gene_attr$`RRST_log_count`), max(gene_attr$standard_log_count)))) +
  scale_y_continuous(limits = c(0, max(max(gene_attr$`RRST_log_count`), max(gene_attr$standard_log_count)))) +
  geom_abline(slope = 1, intercept = 0, color = "red", linetype = "longdash") +
  labs(x = "standard protocol", y = "RRST", title = "log1p(UMI counts)") +
  theme_minimal() +
  theme(axis.text = element_text(size = 11, colour = "black"), axis.title = element_text(size = 14, colour = "black"),
        plot.title = element_text(size = 14, face = "bold", colour = "black"))

p2 <- ggplot(gene_attr, aes_string("standard_det_rate", "RRST_det_rate")) +
  geom_point(size = 0.3) +
  geom_abline(slope = 1, intercept = 0, color = "red", linetype = "longdash") +
  labs(x = "standard protocol", y = "RRST", title = "detection rate") +
  theme_minimal() +
  theme(axis.text = element_text(size = 11, colour = "black"), axis.title = element_text(size = 14, colour = "black"),
        plot.title = element_text(size = 14, face = "bold", colour = "black"))

p1 + p2

```

```{r eval=FALSE}

# Export plots
jpeg(filename = "plots/mousebrain_gene_scatter_counts.jpg", width = 1000, height = 1000, res = 300)
print(p1)
dev.off()
jpeg(filename = "plots/mousebrain_gene_scatter_det_rater.jpg", width = 1000, height = 1000, res = 300)
print(p2)
dev.off()

```

## Supplementary figure - biotype content mouse brain
***

```{r fig.width=10, fig.height=3}

gene_annotations <- read.table(file = "../data/genes/mgenes.tsv", header = TRUE)
rownames(gene_annotations) <- gene_annotations$gene_name

# rename gene type for mitochondrial and ribosomal protein coding genes
rp.genes <- grep(pattern = "^Rpl|^Rps", x = rownames(gene_attr), value = TRUE)
mt.genes <- grep(pattern = "^mt-", x = rownames(gene_attr), value = TRUE)
gene_annotations$gene_type <- ifelse(gene_annotations$gene_name %in% rp.genes, "ribosomal\nprotein_coding", gene_annotations$gene_type)
gene_annotations$gene_type <- ifelse(gene_annotations$gene_name %in% mt.genes, "mitochondrial", gene_annotations$gene_type)

gene_attr$gene_type <- gene_annotations[rownames(gene_attr), "gene_type"]
gene_attr <- gene_attr %>%
  mutate(observed_in = case_when(RRST_count > 0 & standard_count > 0 ~ "both",
                                 RRST_count > 0 & standard_count == 0 ~ "RRST",
                                 RRST_count == 0 & standard_count > 0 ~ "standard")) %>%
  mutate(gene = rownames(.))

lvls <- gene_attr %>% 
           group_by(gene_type) %>% 
           summarize(tot = sum(RRST_count + standard_count)) %>% 
           arrange(-tot) %>% 
           pull(gene_type)

gene_attr_summarized <- gene_attr %>% 
               reshape2::melt(measure.vars = c("RRST_count", "standard_count")) %>%
               filter(value > 0) %>%
               group_by(gene_type, variable) %>% 
               summarize(totTranscript = n(), totCount = sum(value)) %>%
               mutate(gene_type = factor(gene_type, levels = lvls), 
                      variable = ifelse(variable == "RRST_count", "RRST", "standard"))

gene_attr_prop <- gene_attr %>% 
  group_by(gene_type, observed_in) %>%
  summarize(Freq = n()) %>%
  group_by(gene_type) %>%
  mutate(prop = Freq/sum(Freq)) %>%
  mutate(gene_type = factor(gene_type, levels = lvls))

p1 <- ggplot(gene_attr_summarized,
             aes(gene_type, totTranscript, fill = variable)) +
  geom_bar(stat = "identity", position = position_dodge(preserve = 'single')) +
  scale_y_cut(breaks = 100, which = c(1, 2), scales = c(1, 2), space = 0.5) +
  theme_minimal() +
  theme(axis.text.x = element_blank(), axis.title.x = element_blank(), 
        axis.text.y = element_text(size = 12, colour = "black"), 
        axis.title = element_text(size = 14, colour = "black"), 
        legend.text = element_text(size = 14, colour = "black"),
        legend.title = element_text(size = 14, colour = "black")) +
  labs(y = "transcripts detected", fill = "")
p2 <- ggplot(gene_attr_prop, aes("", prop, fill = observed_in)) +
  geom_bar(stat = "identity", color = "black") + 
  coord_polar("y", start = 0) +
  scale_fill_brewer(palette = "Spectral") +
  facet_wrap(~gene_type, strip.position = "bottom", ncol = 10) +
  theme_void() +
  theme(strip.text = element_text(angle = 90, vjust = 0.5, hjust = 1, margin = margin(0.2, 0, 0.2, 0, "cm"), colour = "black", size = 12), 
        legend.text = element_text(size = 14, colour = "black"),
        legend.title = element_text(size = 14, colour = "black"))
p3 <- ggplot(gene_attr_summarized,
               aes(gene_type, totCount, fill = variable)) +
  geom_bar(stat = "identity", position = position_dodge(preserve = 'single')) +
  scale_y_cut(breaks = 1e5, which = c(1, 2), scales = c(1, 2), space = 0.5) +
  theme_minimal() +
  theme(axis.text.x = element_blank(), axis.title.x = element_blank(), 
        axis.text.y = element_text(size = 12, colour = "black"), 
        axis.title = element_text(size = 14, colour = "black"), 
        legend.text = element_text(size = 14, colour = "black"),
        legend.title = element_text(size = 14, colour = "black")) +
  labs(y = "UMI count", fill = "")

p1
p2
p3

```

```{r eval=FALSE}
jpeg("../Suppl_figures/Suppl_Figure_mousebrain_and_prostatecancer/mousebrain_biotype_content1.jpg", width = 2500, height = 500, res = 200, bg = NA)
print(p1)
dev.off()
jpeg("../Suppl_figures/Suppl_Figure_mousebrain_and_prostatecancer/mousebrain_biotype_content2.jpg", width = 2500, height = 500, res = 200, bg = NA)
print(p2)
dev.off()
jpeg("../Suppl_figures/Suppl_Figure_mousebrain_and_prostatecancer/mousebrain_biotype_content3.jpg", width = 2500, height = 500, res = 200, bg = NA)
print(p3)
dev.off()

```

Load prostate cancer data (2xRRST and 2xstandard)

```{r}

PC_infoTable <- subset(infoTable, source == "prostate cancer")
PC <- InputFromTable(infotable = PC_infoTable[c(3:4, 1:2), ])

```

Prostate cancer H&E images

```{r fig.width=8, fig.height=2}

PC <- LoadImages(PC, time.resolve = FALSE)
ImagePlot(PC, ncols = 4)

```

Apply rigid transformations. 

```{r}

# Warp transform
PC <- WarpImages(PC, verbose = TRUE, transforms = list("1" = list(shift.y = -130), "3" = list(angle = 180)))
```

## b - Spatial map of unique genes per spot [prostate cancer]
***

```{r fig.width=8, fig.height=5}

p <- ST.FeaturePlot(PC, features = "nFeature_RNA", ncol = 2, indices = c(3, 2), 
                    label.by = "protocol",  pt.size = 1.2, show.sb = FALSE) &
  theme(plot.title = element_blank(), legend.position = "top", 
        legend.text = element_text(angle = 60, hjust = 1, size = 14, colour = "black"),
        legend.title = element_text(size = 18, face = "bold", colour = "black", vjust = 1),
        strip.text = element_text(size = 18, face = "bold", colour = "black")) &
  labs(fill = "unique genes")

p

```

```{r eval=FALSE}

# Export plot
jpeg(filename = "plots/prostatecancer_unique_genes_spatial.jpg", width = 2200, height = 1400, res = 300)
print(p)
dev.off()

```

## c - Violin plots of unique genes per spot [mousebrain]
***

```{r}

p <- ggplot(PC[[]], aes(arrayID, nFeature_RNA, fill = protocol)) +
  geom_violin(scale = "width") +
  geom_boxplot(width = 0.3) +
  geom_hline(data = PC[[]] %>% dplyr::group_by(protocol) %>% summarize(Median = median(nFeature_RNA)), 
             aes("", yintercept = Median, group = protocol), linetype = "longdash") +
  geom_label(data = PC[[]] %>% dplyr::group_by(protocol) %>% summarize(Median = median(nFeature_RNA)), 
             aes("", y = Median, group = protocol, label = Median)) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1, size = 14, colour = "black"), 
        axis.text.y = element_text(size = 14, colour = "black"), 
        strip.text = element_text(size = 18, face = "bold", colour = "black")) +
  labs(y = "", x = "") +
  facet_grid(~protocol, scales = "free") +
  guides(fill = "none") +
  scale_x_discrete(labels = c("", "Rep1", "Rep2"))

p

```

```{r eval=FALSE}

jpeg(filename = "plots/prostatecancer_unique_genes_violin.jpg", width = 2000, height = 900, res = 300)
print(p)
dev.off()

```

Compute gene attributes

```{r}

gene_attr <- do.call(cbind, lapply(c("RRST", "standard"), function(pro) {
  umis <- GetAssayData(PC, slot = "counts", assay = "RNA")[, PC$protocol %in% pro]
  gene_attr_protocol <- data.frame(rowSums(umis), rowMeans(umis > 0), row.names = rownames(umis))
  gene_attr_protocol <- setNames(gene_attr_protocol, nm = c(paste0(gsub(pattern = " ", replacement = "_", x = pro), "_count"), 
                                                            paste0(gsub(pattern = " ", replacement = "_", x = pro), "_det_rate")))
  return(gene_attr_protocol)
}))

gene_attr <- gene_attr %>%
  mutate("RRST_log_count" = log1p(`RRST_count`), "standard_log_count" = log1p(`standard_count`))

```


Subset genes to include those available in the Visium FFPE probe set.

```{r}

human_probes <- read.csv("../data/sheets/Visium_Human_Transcriptome_Probe_Set_v1.0_GRCh38-2020-A.csv", skip = 5, header = TRUE)
human_probes$gene_name <- do.call(rbind, strsplit(human_probes$probe_id, "\\|"))[, 2]

```

## d - Gene-gene scatter plot of detection rate and UMI counts [prostate cancer]
***


```{r fig.width=8, fig.height=4}

p1 <- ggplot(gene_attr %>% mutate(gene = rownames(.)) %>% subset(gene %in% human_probes$gene_name), aes_string("standard_log_count", "RRST_log_count")) +
  geom_point(size = 0.3) +
  ggpubr::stat_cor() +
  scale_x_continuous(limits = c(0, max(max(gene_attr$`RRST_log_count`), max(gene_attr$standard_log_count)))) +
  scale_y_continuous(limits = c(0, max(max(gene_attr$`RRST_log_count`), max(gene_attr$standard_log_count)))) +
  geom_abline(slope = 1, intercept = 0, color = "red", linetype = "longdash") +
  labs(x = "standard protocol", y = "RRST", title = "log1p(UMI counts)") +
  theme_minimal() +
  theme(axis.text = element_text(size = 11, colour = "black"), axis.title = element_text(size = 14, colour = "black"),
        plot.title = element_text(size = 14, face = "bold", colour = "black"))

p2 <- ggplot(gene_attr, aes_string("standard_det_rate", "RRST_det_rate")) +
  geom_point(size = 0.3) +
  geom_abline(slope = 1, intercept = 0, color = "red", linetype = "longdash") +
  labs(x = "standard protocol", y = "RRST", title = "detection rate") +
  theme_minimal() +
  theme(axis.text = element_text(size = 11, colour = "black"), axis.title = element_text(size = 14, colour = "black"),
        plot.title = element_text(size = 14, face = "bold", colour = "black"))

p1 + p2

```

```{r eval=FALSE}

jpeg(filename = "plots/prostatecancer_gene_scatter_counts.jpg", width = 1000, height = 1000, res = 300)
print(p1)
dev.off()
jpeg(filename = "plots/prostatecancer_gene_scatter_det_rate.jpg", width = 1000, height = 1000, res = 300)
print(p2)
dev.off()

```


## Supplementary figure - biotype content prostate cancer
***

```{r fig.width=10, fig.height=3}

gene_annotations <- read.table(file = "../data/genes/hgenes.tsv", header = TRUE)
rownames(gene_annotations) <- gene_annotations$gene_name

# rename gene type for mitochondrial and ribosomal protein coding genes
rp.genes <- grep(pattern = "^RPL|^RPS", x = rownames(gene_attr), value = TRUE)
mt.genes <- grep(pattern = "^MT-", x = rownames(gene_attr), value = TRUE)
gene_annotations$gene_type <- ifelse(gene_annotations$gene_name %in% rp.genes, "ribosomal\nprotein_coding", gene_annotations$gene_type)
gene_annotations$gene_type <- ifelse(gene_annotations$gene_name %in% mt.genes, "mitochondrial", gene_annotations$gene_type)

gene_attr$gene_type <- gene_annotations[rownames(gene_attr), "gene_type"]
gene_attr <- gene_attr %>%
  mutate(observed_in = case_when(RRST_count > 0 & standard_count > 0 ~ "both",
                                 RRST_count > 0 & standard_count == 0 ~ "RRST",
                                 RRST_count == 0 & standard_count > 0 ~ "standard")) %>%
  mutate(gene = rownames(.))

gene_attr <- subset(gene_attr, !gene_type %in% c("IG_C_pseudogene", "TR_V_pseudogene"))

lvls <- gene_attr %>% 
           group_by(gene_type) %>% 
           summarize(tot = sum(RRST_count + standard_count)) %>% 
           arrange(-tot) %>% 
           pull(gene_type)

gene_attr_summarized <- gene_attr %>% 
               reshape2::melt(measure.vars = c("RRST_count", "standard_count")) %>%
               filter(value > 0) %>%
               group_by(gene_type, variable) %>% 
               summarize(totTranscript = n(), totCount = sum(value)) %>%
               mutate(gene_type = factor(gene_type, levels = lvls), 
                      variable = ifelse(variable == "RRST_count", "RRST", "standard"))

gene_attr_prop <- gene_attr %>% 
  group_by(gene_type, observed_in) %>%
  summarize(Freq = n()) %>%
  group_by(gene_type) %>%
  mutate(prop = Freq/sum(Freq)) %>%
  mutate(gene_type = factor(gene_type, levels = lvls))

p1 <- ggplot(gene_attr_summarized,
             aes(gene_type, totTranscript, fill = variable)) +
  geom_bar(stat = "identity", position = position_dodge(preserve = 'single')) +
  scale_y_cut(breaks = 100, which = c(1, 2), scales = c(1, 2), space = 0.5) +
  theme_minimal() +
  theme(axis.text.x = element_blank(), axis.title.x = element_blank(), 
        axis.text.y = element_text(size = 12, colour = "black"), 
        axis.title = element_text(size = 14, colour = "black"), 
        legend.text = element_text(size = 14, colour = "black"),
        legend.title = element_text(size = 14, colour = "black")) +
  labs(y = "transcripts detected", fill = "")
p2 <- ggplot(gene_attr_prop, aes("", prop, fill = observed_in)) +
  geom_bar(stat = "identity", color = "black") + 
  coord_polar("y", start = 0) +
  scale_fill_brewer(palette = "Spectral") +
  facet_wrap(~gene_type, strip.position = "bottom", ncol = 10) +
  theme_void() +
  theme(strip.text = element_text(angle = 90, vjust = 0.5, hjust = 1, margin = margin(0.2, 0, 0.2, 0, "cm"), colour = "black", size = 12), 
        legend.text = element_text(size = 14, colour = "black"),
        legend.title = element_text(size = 14, colour = "black"))
p3 <- ggplot(gene_attr_summarized,
               aes(gene_type, totCount, fill = variable)) +
  geom_bar(stat = "identity", position = position_dodge(preserve = 'single')) +
  scale_y_cut(breaks = 1e5, which = c(1, 2), scales = c(1, 2), space = 0.5) +
  theme_minimal() +
  theme(axis.text.x = element_blank(), axis.title.x = element_blank(), 
        axis.text.y = element_text(size = 12, colour = "black"), 
        axis.title = element_text(size = 14, colour = "black"), 
        legend.text = element_text(size = 14, colour = "black"),
        legend.title = element_text(size = 14, colour = "black")) +
  labs(y = "UMI count", fill = "")

p1
p2
p3

```

```{r eval=FALSE}

jpeg("../Suppl_figures/Suppl_Figure_mousebrain_and_prostatecancer/prostatecancer_biotype_content1.jpg", width = 2500, height = 500, res = 200, bg = NA)
print(p1)
dev.off()
jpeg("../Suppl_figures/Suppl_Figure_mousebrain_and_prostatecancer/prostatecancer_biotype_content2.jpg", width = 2500, height = 500, res = 200, bg = NA)
print(p2)
dev.off()
jpeg("../Suppl_figures/Suppl_Figure_mousebrain_and_prostatecancer/prostatecancer_biotype_content3.jpg", width = 2500, height = 500, res = 200, bg = NA)
print(p3)
dev.off()

```

# date
***

```{r}

date()

```

# Session
***

```{r}

sessionInfo()

```


